{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 1,
   "id": "6d42ec27",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import pickle\n",
    "from flask import Flask\n",
    "from app import *"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "1e799efb",
   "metadata": {},
   "source": [
    "# Datasets Round 1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "id": "ef5a4815",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [1] Add dataset Wolfram-Schauerte_2022\n",
    "\n",
    "description = 'Exponentially growing E. coli infected with T4 phage (MOI = 3.1) in biological triplicates. Samples taken before (0) and 1, 4, 7, 20 min post-infection and analysed by stranded RNA-seq (n=3). [GSE211026]'\n",
    "doi = 'https://doi.org/10.3390/v14112502'\n",
    "\n",
    "with app.app_context():\n",
    "    addFullDataset(phageName='T4 phage', phageDes = 'Bacteriophage T4', phageID = 'NC_000866.4', phageType = 'virulent',\n",
    "               hostName = 'E. coli B strain', hostDes = 'E. coli K12 MG1655 reference', hostID = 'U00096.3', hostGroup = 'Escherichia coli',\n",
    "               rnaSeqPath = 'Datasets/Wolfram_Schauerte_2022/Notebook/Wolfram-Schauerte_fractional_expression.tsv', norm = 'fractional',\n",
    "               name = 'Wolfram-Schauerte_2022', year = 2022, journal = 'Viruses', firstauthor = 'Wolfram-Schauerte', pubmedID = 36423111, description = description, doi = doi, upload_date = '25-08-12',\n",
    "               phageGenomePath = 'Datasets/Wolfram_Schauerte_2022/Notebook/Wolfram-Schauerte_phage_gff3.tsv', phageGenomeName = 'T4 phage genome',\n",
    "               hostGenomePath = 'Datasets/Wolfram_Schauerte_2022/Notebook/Wolfram-Schauerte_host_gff3.tsv', hostGenomeName = 'E. coli genome')\n",
    "\n",
    "\n",
    "# Add more datasets with different normalizations of Wolfram-Schauerte_2022\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Wolfram_Schauerte_2022/Notebook/Wolfram-Schauerte_full_TPM.tsv',\n",
    "                     norm = 'TPM', name = 'Wolfram-Schauerte_2022', year = 2022, journal = 'Viruses', description = description,\n",
    "                     firstauthor = 'Wolfram-Schauerte', pubmedID = 36423111, host_id=1, phage_id=1, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Wolfram_Schauerte_2022/Notebook/Wolfram-Schauerte_TPM_means.tsv',\n",
    "                     norm = 'TPM_means', name = 'Wolfram-Schauerte_2022', year = 2022, journal = 'Viruses', description = description,\n",
    "                     firstauthor = 'Wolfram-Schauerte', pubmedID = 36423111, host_id=1, phage_id=1, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Wolfram_Schauerte_2022/Notebook/Wolfram-Schauerte_TPM_std.tsv',\n",
    "                     norm = 'TPM_std', name = 'Wolfram-Schauerte_2022', year = 2022, journal = 'Viruses', description = description,\n",
    "                     firstauthor = 'Wolfram-Schauerte', pubmedID = 36423111, host_id=1, phage_id=1, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Wolfram_Schauerte_2022/Notebook/Wolfram-Schauerte_full_raw_counts.tsv',\n",
    "                     norm = 'raw_counts', name = 'Wolfram-Schauerte_2022', year = 2022, journal = 'Viruses', description = description,\n",
    "                     firstauthor = 'Wolfram-Schauerte', pubmedID = 36423111, host_id=1, phage_id=1, doi = doi, upload_date = '25-08-12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "id": "599c2cb7",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [2] Add T4 minus toxIN datasets from Guegler 2021 referring to same phage and host as Wolfram-Schauerte 2022\n",
    "\n",
    "description = 'E. coli not expressing toxIN (pBR322 empty vector; -toxIN) infected with T4 phage (MOI = 5) in biological duplicates. Samples taken before (0) and 2.5, 5, 10, 20, 30 min post-infection (n=2). [GSE161794]'\n",
    "doi = 'https://doi.org/10.1016/j.molcel.2021.03.027'\n",
    "\n",
    "with app.app_context():\n",
    "    hostAdd = Host(name = 'E. coli K 12 pBR322 empty', group = 'Escherichia coli', description = 'E. coli K12 MG1655 without toxIN expression', ncbi_id = 'U00096.3')\n",
    "    db.session.add(hostAdd)\n",
    "    db.session.commit()\n",
    "\n",
    "with app.app_context():\n",
    "    hostgff = pd.read_csv('Datasets/Guegler_2021/Notebook_T4/Guegler_T4_minusToxIN_host_gff3.tsv', sep='\\t', comment = '#')\n",
    "    hostgenome = pickle.dumps(hostgff)\n",
    "    hostGFF = HostGenome(name = 'E. coli K12 MG1655 genome', host_id = 2, gff_data = hostgenome)\n",
    "    db.session.add(hostGFF)\n",
    "    db.session.commit()\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Guegler_2021/Notebook_T4/Guegler_T4_minusToxIN_fractional_expression.tsv',\n",
    "                     norm = 'fractional', name = 'Guegler_2021_T4_minus_toxIN', year = 2021, journal = 'Molecular Cell', description = description,\n",
    "                     firstauthor = 'Guegler', pubmedID = 33838104, host_id=2, phage_id=1, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Guegler_2021/Notebook_T4/Guegler_T4_minusToxIN_full_TPM.tsv',\n",
    "                     norm = 'TPM', name = 'Guegler_2021_T4_minus_toxIN', year = 2021, journal = 'Molecular Cell', description = description,\n",
    "                     firstauthor = 'Guegler', pubmedID = 33838104, host_id=2, phage_id=1, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Guegler_2021/Notebook_T4/Guegler_T4_minusToxIN_TPM_means.tsv',\n",
    "                     norm = 'TPM_means', name = 'Guegler_2021_T4_minus_toxIN', year = 2021, journal = 'Molecular Cell', description = description,\n",
    "                     firstauthor = 'Guegler', pubmedID = 33838104, host_id=2, phage_id=1, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Guegler_2021/Notebook_T4/Guegler_T4_minusToxIN_TPM_std.tsv',\n",
    "                     norm = 'TPM_std', name = 'Guegler_2021_T4_minus_toxIN', year = 2021, journal = 'Molecular Cell', description = description,\n",
    "                     firstauthor = 'Guegler', pubmedID = 33838104, host_id=2, phage_id=1, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Guegler_2021/Notebook_T4/Guegler_T4_minusToxIN_full_raw_counts.tsv',\n",
    "                     norm = 'raw_counts', name = 'Guegler_2021_T4_minus_toxIN', year = 2021, journal = 'Molecular Cell', description = description,\n",
    "                     firstauthor = 'Guegler', pubmedID = 33838104, host_id=2, phage_id=1, doi = doi, upload_date = '25-08-12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "fb3cf3d5",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [3] Add T4 plustoxIN datasets from Guegler 2021 referring to same phage and host as Wolfram-Schauerte 2022\n",
    "\n",
    "description = 'E. coli induced to express toxIN (pBR322-toxIN) infected with T4 phage (MOI = 5) in biological duplicates. Samples taken before (0) and 2.5, 5, 10, 20, 30 min post-infection (n=2). [GSE161794]'\n",
    "doi = 'https://doi.org/10.1016/j.molcel.2021.03.027'\n",
    "\n",
    "with app.app_context():\n",
    "    hostAdd = Host(name = 'E. coli K 12 pBR322-toxIN', group = 'Escherichia coli', description = 'E. coli K12 MG1655 expressing toxIN', ncbi_id = 'U00096.3')\n",
    "    db.session.add(hostAdd)\n",
    "    db.session.commit()\n",
    "\n",
    "with app.app_context():\n",
    "    hostgff = pd.read_csv('Datasets/Guegler_2021/Notebook_T4/Guegler_T4_plusToxIN_host_gff3.tsv', sep='\\t', comment = '#')\n",
    "    hostgenome = pickle.dumps(hostgff)\n",
    "    hostGFF = HostGenome(name = 'E. coli K12 MG1655 genome', host_id = 3, gff_data = hostgenome)\n",
    "    db.session.add(hostGFF)\n",
    "    db.session.commit()\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Guegler_2021/Notebook_T4/Guegler_T4_plusToxIN_fractional_expression.tsv',\n",
    "                     norm = 'fractional', name = 'Guegler_2021_T4_plus_toxIN', year = 2021, journal = 'Molecular Cell', description = description,\n",
    "                     firstauthor = 'Guegler', pubmedID = 33838104, host_id=3, phage_id=1, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Guegler_2021/Notebook_T4/Guegler_T4_plusToxIN_full_TPM.tsv',\n",
    "                     norm = 'TPM', name = 'Guegler_2021_T4_plus_toxIN', year = 2021, journal = 'Molecular Cell', description = description,\n",
    "                     firstauthor = 'Guegler', pubmedID = 33838104, host_id=3, phage_id=1, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Guegler_2021/Notebook_T4/Guegler_T4_plusToxIN_TPM_means.tsv',\n",
    "                     norm = 'TPM_means', name = 'Guegler_2021_T4_plus_toxIN', year = 2021, journal = 'Molecular Cell', description = description,\n",
    "                     firstauthor = 'Guegler', pubmedID = 33838104, host_id=3, phage_id=1, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Guegler_2021/Notebook_T4/Guegler_T4_plusToxIN_TPM_std.tsv',\n",
    "                     norm = 'TPM_std', name = 'Guegler_2021_T4_plus_toxIN', year = 2021, journal = 'Molecular Cell', description = description,\n",
    "                     firstauthor = 'Guegler', pubmedID = 33838104, host_id=3, phage_id=1, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Guegler_2021/Notebook_T4/Guegler_T4_plusToxIN_full_raw_counts.tsv',\n",
    "                     norm = 'raw_counts', name = 'Guegler_2021_T4_plus_toxIN', year = 2021, journal = 'Molecular Cell', description = description,\n",
    "                     firstauthor = 'Guegler', pubmedID = 33838104, host_id=3, phage_id=1, doi = doi, upload_date = '25-08-12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "id": "fdc0face",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [4] Add T7 minus toxIN datasets from Guegler 2021 referring to same host as Guegler_T4_minusToxIN\n",
    "\n",
    "description = 'E. coli not expressing toxIN (pBR322 empty vector; -toxIN) infected with T7 phage (MOI = 5) in biological duplicates. Samples taken before (0) and 2.5, 5, 10, 20, 30 min post-infection (n=2). [GSE161795]'\n",
    "doi = 'https://doi.org/10.1016/j.molcel.2021.03.027'\n",
    "\n",
    "with app.app_context():\n",
    "    phageAdd = Phage(name = 'T7 phage', description = 'Bacteriophage T7', ncbi_id = 'NC_001604.1', phage_type = 'virulent')\n",
    "    db.session.add(phageAdd)\n",
    "    db.session.commit()\n",
    "\n",
    "with app.app_context():\n",
    "    phagegff = pd.read_csv('Datasets/Guegler_2021/Notebook_T7/Guegler_T7_minusToxIN_phage_gff3.tsv', sep='\\t', comment = '#')\n",
    "    phagegenome = pickle.dumps(phagegff)\n",
    "    phageGFF = PhageGenome(name = 'T7 phage genome', phage_id = 2, gff_data = phagegenome)\n",
    "    db.session.add(phageGFF)\n",
    "    db.session.commit()\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Guegler_2021/Notebook_T7/Guegler_T7_minusToxIN_fractional_expression.tsv',\n",
    "                     norm = 'fractional', name = 'Guegler_2021_T7_minus_toxIN', year = 2021, journal = 'Molecular Cell', description = description,\n",
    "                     firstauthor = 'Guegler', pubmedID = 33838104, host_id=2, phage_id=2, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Guegler_2021/Notebook_T7/Guegler_T7_plusToxIN_full_TPM.tsv',\n",
    "                     norm = 'TPM', name = 'Guegler_2021_T7_minus_toxIN', year = 2021, journal = 'Molecular Cell', description = description,\n",
    "                     firstauthor = 'Guegler', pubmedID = 33838104, host_id=2, phage_id=2, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Guegler_2021/Notebook_T7/Guegler_T7_plusToxIN_TPM_means.tsv',\n",
    "                     norm = 'TPM_means', name = 'Guegler_2021_T7_minus_toxIN', year = 2021, journal = 'Molecular Cell', description = description,\n",
    "                     firstauthor = 'Guegler', pubmedID = 33838104, host_id=2, phage_id=2, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Guegler_2021/Notebook_T7/Guegler_T7_plusToxIN_TPM_std.tsv',\n",
    "                     norm = 'TPM_std', name = 'Guegler_2021_T7_minus_toxIN', year = 2021, journal = 'Molecular Cell', description = description,\n",
    "                     firstauthor = 'Guegler', pubmedID = 33838104, host_id=2, phage_id=2, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Guegler_2021/Notebook_T7/Guegler_T7_plusToxIN_full_raw_counts.tsv',\n",
    "                     norm = 'raw_counts', name = 'Guegler_2021_T7_minus_toxIN', year = 2021, journal = 'Molecular Cell', description = description,\n",
    "                     firstauthor = 'Guegler', pubmedID = 33838104, host_id=2, phage_id=2, doi = doi, upload_date = '25-08-12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "e6963935",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [5] Add T7 plus toxIN datasets from Guegler 2021 referring to same host as Guegler_T4_plusToxIN\n",
    "\n",
    "description = 'E. coli induced to express toxIN (pBR322-toxIN) infected with T7 phage (MOI = 5) in biological duplicates. Samples taken before (0) and 2.5, 5, 10, 20, 30 min post-infection (n=2). [GSE161795]'\n",
    "doi = 'https://doi.org/10.1016/j.molcel.2021.03.027'\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Guegler_2021/Notebook_T7/Guegler_T7_plusToxIN_fractional_expression.tsv',\n",
    "                     norm = 'fractional', name = 'Guegler_2021_T7_plus_toxIN', year = 2021, journal = 'Molecular Cell', description = description,\n",
    "                     firstauthor = 'Guegler', pubmedID = 33838104, host_id=3, phage_id=2, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Guegler_2021/Notebook_T7/Guegler_T7_plusToxIN_full_TPM.tsv',\n",
    "                     norm = 'TPM', name = 'Guegler_2021_T7_plus_toxIN', year = 2021, journal = 'Molecular Cell', description = description,\n",
    "                     firstauthor = 'Guegler', pubmedID = 33838104, host_id=3, phage_id=2, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Guegler_2021/Notebook_T7/Guegler_T7_plusToxIN_TPM_means.tsv',\n",
    "                     norm = 'TPM_means', name = 'Guegler_2021_T7_plus_toxIN', year = 2021, journal = 'Molecular Cell', description = description,\n",
    "                     firstauthor = 'Guegler', pubmedID = 33838104, host_id=3, phage_id=2, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Guegler_2021/Notebook_T7/Guegler_T7_plusToxIN_TPM_std.tsv',\n",
    "                     norm = 'TPM_std', name = 'Guegler_2021_T7_plus_toxIN', year = 2021, journal = 'Molecular Cell', description = description,\n",
    "                     firstauthor = 'Guegler', pubmedID = 33838104, host_id=3, phage_id=2, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Guegler_2021/Notebook_T7/Guegler_T7_plusToxIN_full_raw_counts.tsv',\n",
    "                     norm = 'raw_counts', name = 'Guegler_2021_T7_plus_toxIN', year = 2021, journal = 'Molecular Cell', description = description,\n",
    "                     firstauthor = 'Guegler', pubmedID = 33838104, host_id=3, phage_id=2, doi = doi, upload_date = '25-08-12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "00412514",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [6] Add datasets Finstrlova_2022 Newman\n",
    "\n",
    "description = 'Infection of S. aureus strain Newman with Staphylococcus virus K (MOI = 7) sampled in biological triplicates before (0) and 2, 5, 10, 20, 30 minutes post-infection (n=3). [GSE190637]'\n",
    "doi = 'https://doi.org/10.1128/spectrum.00123-22'\n",
    "\n",
    "with app.app_context():\n",
    "    addFullDataset(phageName='Kayvirus phage K', phageDes = 'Staphylococcus virus K', phageID = 'NC_005880.2', phageType = 'virulent',\n",
    "               hostName = 'S. aureus strain Newman', hostDes = 'Staphylococcus aureus strain Newman', hostID = 'NC_009641.1', hostGroup = 'Staphylococcus aureus',\n",
    "               rnaSeqPath = 'Datasets/Finstrlova_2022/Notebook_Newman/Finstrlova_Newman_fractional_expression.tsv', norm = 'fractional',\n",
    "               name = 'Finstrlova_2022_Newman', year = 2022, journal = 'Microbiology Spectrum', firstauthor = 'Finstrlova', pubmedID = 3545752, description = description, doi = doi, upload_date = '25-08-12',\n",
    "               phageGenomePath = 'Datasets/Finstrlova_2022/Notebook_Newman/Finstrlova_Newman_phage_gff3.tsv', phageGenomeName = 'K virus genome',\n",
    "               hostGenomePath = 'Datasets/Finstrlova_2022/Notebook_Newman/Finstrlova_Newman_host_gff3.tsv', hostGenomeName = 'S. aureus Newman genome')\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Finstrlova_2022/Notebook_Newman/Finstrlova_Newman_full_TPM.tsv', norm = 'TPM',\n",
    "               name = 'Finstrlova_2022_Newman', year = 2022, journal = 'Microbiology Spectrum', firstauthor = 'Finstrlova', pubmedID = 3545752, description = description, host_id=4, phage_id=3, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Finstrlova_2022/Notebook_Newman/Finstrlova_Newman_TPM_means.tsv', norm = 'TPM_means',\n",
    "               name = 'Finstrlova_2022_Newman', year = 2022, journal = 'Microbiology Spectrum', firstauthor = 'Finstrlova', pubmedID = 3545752, description = description, host_id=4, phage_id=3, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Finstrlova_2022/Notebook_Newman/Finstrlova_Newman_TPM_std.tsv', norm = 'TPM_std',\n",
    "               name = 'Finstrlova_2022_Newman', year = 2022, journal = 'Microbiology Spectrum', firstauthor = 'Finstrlova', pubmedID = 3545752, description = description, host_id=4, phage_id=3, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Finstrlova_2022/Notebook_Newman/Finstrlova_Newman_full_raw_counts.tsv', norm = 'raw_counts',\n",
    "               name = 'Finstrlova_2022_Newman', year = 2022, journal = 'Microbiology Spectrum', firstauthor = 'Finstrlova', pubmedID = 3545752, description = description, host_id=4, phage_id=3, doi = doi, upload_date = '25-08-12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "63727834",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [7] Add datasets Finstrlova_2022 SH1000 \n",
    "\n",
    "description = 'Infection of S. aureus strain SH1000 with Staphylococcus virus K (MOI = 7) sampled in biological triplicates before (0) and 2, 5, 10, 20, 30 minutes post-infection (n=3). [GSE190637]'\n",
    "doi = 'https://doi.org/10.1128/spectrum.00123-22'\n",
    "\n",
    "with app.app_context():\n",
    "    hostAdd = Host(name = 'S. aureus SH1000', description = 'Staphylococcus aureus strain SH1000', ncbi_id = 'NZ_CP059180.1', group = 'Staphylococcus aureus')\n",
    "    db.session.add(hostAdd)\n",
    "    db.session.commit()\n",
    "\n",
    "with app.app_context():\n",
    "    hostgff = pd.read_csv('Datasets/Finstrlova_2022/Notebook_SH1000/Finstrlova_SH1000_host_gff3.tsv', sep='\\t', comment = '#')\n",
    "    hostgenome = pickle.dumps(hostgff)\n",
    "    hostGFF = HostGenome(name = 'S. aureus SH1000 genome', host_id = 5, gff_data = hostgenome)\n",
    "    db.session.add(hostGFF)\n",
    "    db.session.commit()\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Finstrlova_2022/Notebook_SH1000/Finstrlova_SH1000_fractional_expression.tsv', norm = 'fractional',\n",
    "               name = 'Finstrlova_2022_SH1000', year = 2022, journal = 'Microbiology Spectrum', firstauthor = 'Finstrlova', pubmedID = 3545752, description = description, host_id=5, phage_id=3, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Finstrlova_2022/Notebook_SH1000/Finstrlova_SH1000_full_TPM.tsv', norm = 'TPM',\n",
    "               name = 'Finstrlova_2022_SH1000', year = 2022, journal = 'Microbiology Spectrum', firstauthor = 'Finstrlova', pubmedID = 3545752, description = description, host_id=5, phage_id=3, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Finstrlova_2022/Notebook_SH1000/Finstrlova_SH1000_TPM_means.tsv', norm = 'TPM_means',\n",
    "               name = 'Finstrlova_2022_SH1000', year = 2022, journal = 'Microbiology Spectrum', firstauthor = 'Finstrlova', pubmedID = 3545752, description = description, host_id=5, phage_id=3, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Finstrlova_2022/Notebook_SH1000/Finstrlova_SH1000_TPM_std.tsv', norm = 'TPM_std',\n",
    "               name = 'Finstrlova_2022_SH1000', year = 2022, journal = 'Microbiology Spectrum', firstauthor = 'Finstrlova', pubmedID = 3545752, description = description, host_id=5, phage_id=3, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Finstrlova_2022/Notebook_SH1000/Finstrlova_SH1000_full_raw_counts.tsv', norm = 'raw_counts',\n",
    "               name = 'Finstrlova_2022_SH1000', year = 2022, journal = 'Microbiology Spectrum', firstauthor = 'Finstrlova', pubmedID = 3545752, description = description, host_id=5, phage_id=3, doi = doi, upload_date = '25-08-12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "a5012622",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [8] Add datasets Kuptsov_2022\n",
    "\n",
    "description = 'Infection of S. aureus strain SA515 with Staphylococcus phage vB_SauM-515A1 (MOI = 10) sampled in biological triplicates at 5, 15, 30 minutes post-infection (n=3). [PRJNA659920]'\n",
    "doi = 'https://doi.org/10.3390/v14030567'\n",
    "\n",
    "with app.app_context():\n",
    "    addFullDataset(phageName='vB_SauM-515A1 kayvirus', phageDes = 'Staphylococcus phage vB_SauM-515A1', phageID = 'MN047438.1', phageType = 'virulent',\n",
    "               hostName = 'S. aureus strain SA515', hostDes = 'Staphylococcus aureus strain SA515', hostID = 'GCA_022352045.1', hostGroup = 'Staphylococcus aureus',\n",
    "               rnaSeqPath = 'Datasets/Kuptsov_2022/Notebook/Kuptsov_fractional_expression.tsv', norm = 'fractional',\n",
    "               name = 'Kuptsov_2022', year = 2022, journal = 'Viruses', firstauthor = 'Kuptsov', pubmedID = 35336974, description = description, doi = doi, upload_date = '25-08-12',\n",
    "               phageGenomePath = 'Datasets/Kuptsov_2022/Notebook/Kuptsov_phage_gff3.tsv', phageGenomeName = 'vB_SauM-515A1 virus genome',\n",
    "               hostGenomePath = 'Datasets/Kuptsov_2022/Notebook/Kuptsov_host_gff3.tsv', hostGenomeName = 'S. aureus SA515 genome')\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Kuptsov_2022/Notebook/Kuptsov_full_TPM.tsv', norm = 'TPM',\n",
    "               name = 'Kuptsov_2022', year = 2022, journal = 'Viruses', firstauthor = 'Kuptsov', pubmedID = 35336974, description = description, host_id=6, phage_id=4, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Kuptsov_2022/Notebook/Kuptsov_TPM_means.tsv', norm = 'TPM_means',\n",
    "               name = 'Kuptsov_2022', year = 2022, journal = 'Viruses', firstauthor = 'Kuptsov', pubmedID = 35336974, description = description, host_id=6, phage_id=4, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Kuptsov_2022/Notebook/Kuptsov_TPM_std.tsv', norm = 'TPM_std',\n",
    "               name = 'Kuptsov_2022', year = 2022, journal = 'Viruses', firstauthor = 'Kuptsov', pubmedID = 35336974, description = description, host_id=6, phage_id=4, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Kuptsov_2022/Notebook/Kuptsov_full_raw_counts.tsv', norm = 'raw_counts',\n",
    "               name = 'Kuptsov_2022', year = 2022, journal = 'Viruses', firstauthor = 'Kuptsov', pubmedID = 35336974, description = description, host_id=6, phage_id=4, doi = doi, upload_date = '25-08-12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "id": "820bd49d",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [9] Add datasets Leskinen_2016\n",
    "\n",
    "description = 'Fresh cultures (logarithmic phase) of Yersinia enterocolitica YeO3-R1 were infected with bacteriophage R1-37 (MOI = 10). Samples were taken before (0) and 2, 5, 10, 15, 21, 28, 35, 42, 49 min post-infection (n=1). [GSE77068]'\n",
    "doi = 'https://doi.org/https://doi.org/10.3390/v8040111'\n",
    "\n",
    "with app.app_context():\n",
    "    addFullDataset(phageName='PhiR1-37', phageDes = 'Yersinia phage phiR1-37', phageID = 'AJ972879.2', phageType = 'virulent',\n",
    "               hostName = 'Y. enterocolitica O:3', hostDes = 'Yersinia enterocolitica O:3', hostID = 'NC_017564.1 + NC_017565.1', hostGroup = 'Yersinia enterolitica',\n",
    "               rnaSeqPath = 'Datasets/Leskinen_2016/Notebook/Leskinen_fractional_expression.tsv', norm = 'fractional',\n",
    "               name = 'Leskinen_2016', year = 2016, journal = 'Viruses', firstauthor = 'Leskinen', pubmedID = 27110815, description = description, doi = doi, upload_date = '25-08-12',\n",
    "               phageGenomePath = 'Datasets/Leskinen_2016/Notebook/Leskinen_phage_gff3.tsv', phageGenomeName = 'PhiR1-37 virus genome',\n",
    "               hostGenomePath = 'Datasets/Leskinen_2016/Notebook/Leskinen_host_gff3.tsv', hostGenomeName = 'Y. enterocolitica O:3 genome')\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Leskinen_2016/Notebook/Leskinen_full_TPM.tsv', norm = 'TPM',\n",
    "               name = 'Leskinen_2016', year = 2016, journal = 'Viruses', firstauthor = 'Leskinen', pubmedID = 27110815, description = description, host_id=7, phage_id=5, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Leskinen_2016/Notebook/Leskinen_TPM_means.tsv', norm = 'TPM_means',\n",
    "               name = 'Leskinen_2016', year = 2016, journal = 'Viruses', firstauthor = 'Leskinen', pubmedID = 27110815, description = description, host_id=7, phage_id=5, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Leskinen_2016/Notebook/Leskinen_TPM_std.tsv', norm = 'TPM_std',\n",
    "               name = 'Leskinen_2016', year = 2016, journal = 'Viruses', firstauthor = 'Leskinen', pubmedID = 27110815, description = description, host_id=7, phage_id=5, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Leskinen_2016/Notebook/Leskinen_full_raw_counts.tsv', norm = 'raw_counts',\n",
    "               name = 'Leskinen_2016', year = 2016, journal = 'Viruses', firstauthor = 'Leskinen', pubmedID = 27110815, description = description, host_id=7, phage_id=5, doi = doi, upload_date = '25-08-12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "fec76b76",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [10] Add datasets brandao_2021 in LB medium\n",
    "\n",
    "description = 'P. aeruginosa PAO1 infected with LUZ19 phage in LB medium (MOI = 75). Samples taken before (0) and 5, 10, 15 min post-infection in biological triplicates (n=3). [GSE162278]'\n",
    "doi = 'https://doi.org/10.1080/15476286.2020.1870844'\n",
    "\n",
    "with app.app_context():\n",
    "    addFullDataset(phageName='LUZ19', phageDes = 'Pseudomonas phage LUZ19', phageID = 'NC_010326.1', phageType = 'virulent',\n",
    "               hostName = 'P. aeruginosa PAO1', hostDes = 'Pseudomonas aeruginosa PAO1', hostID = 'NC_002516.2', hostGroup = 'Pseudomonas aeruginosa',\n",
    "               rnaSeqPath = 'Datasets/Brandão_2021/Notebook/Brandao_LB_fractional_expression.tsv', norm = 'fractional',\n",
    "               name = 'Brandão_2021_LB', year = 2021, journal = 'RNA Biology', firstauthor = 'Brandão', pubmedID = 33448239, description = description, doi = doi, upload_date = '25-08-12',\n",
    "               phageGenomePath = 'Datasets/Brandão_2021/Notebook/Brandao_LB_phage_gff3.tsv', phageGenomeName = 'LUZ19 virus genome',\n",
    "               hostGenomePath = 'Datasets/Brandão_2021/Notebook/Brandao_LB_host_gff3.tsv', hostGenomeName = 'P. aeruginosa PAO1 genome')\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Brandão_2021/Notebook/Brandao_LB_full_TPM.tsv', norm = 'TPM',\n",
    "               name = 'Brandão_2021_LB', year = 2021, journal = 'RNA Biology', firstauthor = 'Brandão', pubmedID = 33448239, description = description, host_id=8, phage_id=6, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Brandão_2021/Notebook/Brandao_LB_TPM_means.tsv', norm = 'TPM_means',\n",
    "               name = 'Brandão_2021_LB', year = 2021, journal = 'RNA Biology', firstauthor = 'Brandão', pubmedID = 33448239, description = description, host_id=8, phage_id=6, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Brandão_2021/Notebook/Brandao_LB_TPM_std.tsv', norm = 'TPM_std',\n",
    "               name = 'Brandão_2021_LB', year = 2021, journal = 'RNA Biology', firstauthor = 'Brandão', pubmedID = 33448239, description = description, host_id=8, phage_id=6, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Brandão_2021/Notebook/Brandao_LB_full_raw_counts.tsv', norm = 'raw_counts',\n",
    "               name = 'Brandão_2021_LB', year = 2021, journal = 'RNA Biology', firstauthor = 'Brandão', pubmedID = 33448239, description = description, host_id=8, phage_id=6, doi = doi, upload_date = '25-08-12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "id": "963cf3d3",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [11] Add datasets brandao_2021 in MCCM medium\n",
    "\n",
    "description = 'P. aeruginosa PAO1 infected with LUZ19 phage in mammalian cell culture medium (MCCM) (MOI = 75). Samples taken before (0) and 5, 10, 15 min post-infection in biological triplicates (n=3). [GSE162278]'\n",
    "doi = 'https://doi.org/10.1080/15476286.2020.1870844'\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Brandão_2021/Notebook/Brandao_MCCM_fractional_expression.tsv', norm = 'fractional',\n",
    "               name = 'Brandão_2021_MCCM', year = 2021, journal = 'RNA Biology', firstauthor = 'Brandão', pubmedID = 33448239, description = description, host_id=8, phage_id=6, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Brandão_2021/Notebook/Brandao_MCCM_full_TPM.tsv', norm = 'TPM',\n",
    "               name = 'Brandão_2021_MCCM', year = 2021, journal = 'RNA Biology', firstauthor = 'Brandão', pubmedID = 33448239, description = description, host_id=8, phage_id=6, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Brandão_2021/Notebook/Brandao_MCCM_TPM_means.tsv', norm = 'TPM_means',\n",
    "               name = 'Brandão_2021_MCCM', year = 2021, journal = 'RNA Biology', firstauthor = 'Brandão', pubmedID = 33448239, description = description, host_id=8, phage_id=6, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Brandão_2021/Notebook/Brandao_MCCM_TPM_std.tsv', norm = 'TPM_std',\n",
    "               name = 'Brandão_2021_MCCM', year = 2021, journal = 'RNA Biology', firstauthor = 'Brandão', pubmedID = 33448239, description = description, host_id=8, phage_id=6, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Brandão_2021/Notebook/Brandao_MCCM_full_raw_counts.tsv', norm = 'raw_counts',\n",
    "               name = 'Brandão_2021_MCCM', year = 2021, journal = 'RNA Biology', firstauthor = 'Brandão', pubmedID = 33448239, description = description, host_id=8, phage_id=6, doi = doi, upload_date = '25-08-12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "id": "d38a6da4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [12] Add datasets sprenger_2024 Vibrio Cholerae WT and VP882 WT\n",
    "\n",
    "description = 'Lytic activation of Vibrio Phage VP882 WT in Vibrio Cholerae O1 el tor WT. Samples taken before (0) and 30, 60 min post-infection in biological triplicates (n=3). [GSE247769]'\n",
    "doi = 'https://doi.org/10.1016/j.chom.2024.03.010'\n",
    "\n",
    "with app.app_context():\n",
    "    addFullDataset(phageName='VP882 WT', phageDes = 'Vibrio Phage VP882 WT', phageID = 'NC_009016.1', phageType = 'temperate',\n",
    "               hostName = 'V. Cholerae O1 WT', hostDes = 'Vibrio Cholerae O1 El Tor strain C6706 WT', hostID = 'NC_002505.1 + NC_002506.1', hostGroup = 'Vibrio Cholerae',\n",
    "               rnaSeqPath = 'Datasets/Sprenger_2024/Notebook_1/Sprenger_VP882_WT_fractional_expression.tsv', norm = 'fractional',\n",
    "               name = 'Sprenger_2024_VC_WT_VP882_WT', year = 2024, journal = 'CellPress', firstauthor = 'Sprenger', pubmedID = 38579715, description = description, doi = doi, upload_date = '25-08-12',\n",
    "               phageGenomePath = 'Datasets/Sprenger_2024/Notebook_1/Sprenger_VP882_WT_phage_gff3.tsv', phageGenomeName = 'VP882 genome',\n",
    "               hostGenomePath = 'Datasets/Sprenger_2024/Notebook_1/Sprenger_VP882_WT_host_gff3.tsv', hostGenomeName = 'V. cholerae O1 WT genome')\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Sprenger_2024/Notebook_1/Sprenger_VP882_WT_full_TPM.tsv', norm = 'TPM',\n",
    "               name = 'Sprenger_2024_VC_WT_VP882_WT', year = 2024, journal = 'CellPress', firstauthor = 'Sprenger', pubmedID = 38579715, description = description, host_id=9, phage_id=7, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Sprenger_2024/Notebook_1/Sprenger_VP882_WT_TPM_means.tsv', norm = 'TPM_means',\n",
    "               name = 'Sprenger_2024_VC_WT_VP882_WT', year = 2024, journal = 'CellPress', firstauthor = 'Sprenger', pubmedID = 38579715, description = description, host_id=9, phage_id=7, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Sprenger_2024/Notebook_1/Sprenger_VP882_WT_TPM_std.tsv', norm = 'TPM_std',\n",
    "               name = 'Sprenger_2024_VC_WT_VP882_WT', year = 2024, journal = 'CellPress', firstauthor = 'Sprenger', pubmedID = 38579715, description = description, host_id=9, phage_id=7, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Sprenger_2024/Notebook_1/Sprenger_VP882_WT_full_raw_counts.tsv', norm = 'raw_counts',\n",
    "               name = 'Sprenger_2024_VC_WT_VP882_WT', year = 2024, journal = 'CellPress', firstauthor = 'Sprenger', pubmedID = 38579715, description = description, host_id=9, phage_id=7, doi = doi, upload_date = '25-08-12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "id": "5d3ca94a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [13] Add datasets sprenger_2024 Vibrio Cholerae WT and VP882 delta vpdS\n",
    "\n",
    "description = 'Lytic activation of Vibrio Phage VP882 delta vpdS in Vibrio Cholerae O1 el tor WT. Samples were taken before (0) and 30, 60 min post-infection in biological triplicates. [GSE247769]'\n",
    "doi = 'https://doi.org/10.1016/j.chom.2024.03.010'\n",
    "\n",
    "with app.app_context():\n",
    "    phageAdd = Phage(name = 'VP882 delta vpdS', description = 'Vibrio Phage VP882 delta vpdS', ncbi_id = 'NC_009016.1', phage_type = 'temperate')\n",
    "    db.session.add(phageAdd)\n",
    "    db.session.commit()\n",
    "\n",
    "with app.app_context():\n",
    "    phagegff = pd.read_csv('Datasets/Sprenger_2024/Notebook_1/Sprenger_VP882_delta_vpdS_phage_gff3.tsv', sep='\\t', comment = '#')\n",
    "    phagegenome = pickle.dumps(phagegff)\n",
    "    phageGFF = PhageGenome(name = 'VP882 delta vpdS genome', phage_id = 8, gff_data = phagegenome)\n",
    "    db.session.add(phageGFF)\n",
    "    db.session.commit()\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Sprenger_2024/Notebook_1/Sprenger_VP882_delta_vpdS_fractional_expression.tsv', norm = 'fractional',\n",
    "               name = 'Sprenger_2024_VC_WT_VP882_delta_vpdS', year = 2024, journal = 'CellPress', firstauthor = 'Sprenger', pubmedID = 38579715, description = description, host_id=9, phage_id=8, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Sprenger_2024/Notebook_1/Sprenger_VP882_delta_vpdS_full_TPM.tsv', norm = 'TPM',\n",
    "               name = 'Sprenger_2024_VC_WT_VP882_delta_vpdS', year = 2024, journal = 'CellPress', firstauthor = 'Sprenger', pubmedID = 38579715, description = description, host_id=9, phage_id=8, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Sprenger_2024/Notebook_1/Sprenger_VP882_delta_vpdS_TPM_means.tsv', norm = 'TPM_means',\n",
    "               name = 'Sprenger_2024_VC_WT_VP882_delta_vpdS', year = 2024, journal = 'CellPress', firstauthor = 'Sprenger', pubmedID = 38579715, description = description, host_id=9, phage_id=8, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Sprenger_2024/Notebook_1/Sprenger_VP882_delta_vpdS_TPM_std.tsv', norm = 'TPM_std',\n",
    "               name = 'Sprenger_2024_VC_WT_VP882_delta_vpdS', year = 2024, journal = 'CellPress', firstauthor = 'Sprenger', pubmedID = 38579715, description = description, host_id=9, phage_id=8, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Sprenger_2024/Notebook_1/Sprenger_VP882_delta_vpdS_full_raw_counts.tsv', norm = 'raw_counts',\n",
    "               name = 'Sprenger_2024_VC_WT_VP882_delta_vpdS', year = 2024, journal = 'CellPress', firstauthor = 'Sprenger', pubmedID = 38579715, description = description, host_id=9, phage_id=8, doi = doi, upload_date = '25-08-12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "id": "5f8abe6f",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [14] Add datasets sprenger_2024 Vibrio Cholerae delta tdh and VP882 WT\n",
    "\n",
    "description = 'Lytic activation of Vibrio Phage VP882 WT in Vibrio Cholerae O1 el tor delta tdh. Samples were taken before (0) and 15, 60, 120 min post-infection in biological triplicates. [GSE247770]'\n",
    "doi = 'https://doi.org/10.1016/j.chom.2024.03.010'\n",
    "\n",
    "with app.app_context():\n",
    "    hostAdd = Host(name = 'V. Cholerae O1 delta tdh', description = 'Vibrio Cholerae O1 El Tor strain C6706 delta tdh', ncbi_id = 'NC_009016.1', group = 'Vibrio Cholerae')\n",
    "    db.session.add(hostAdd)\n",
    "    db.session.commit()\n",
    "\n",
    "with app.app_context():\n",
    "    hostgff = pd.read_csv('Datasets/Sprenger_2024/Notebook_2/Sprenger_VC_delta_tdh_host_gff3.tsv', sep='\\t', comment = '#')\n",
    "    hostgenome = pickle.dumps(hostgff)\n",
    "    hostGFF = HostGenome(name = 'V. cholerae O1 delta tdh genome', host_id = 10, gff_data = hostgenome)\n",
    "    db.session.add(hostGFF)\n",
    "    db.session.commit()\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Sprenger_2024/Notebook_2/Sprenger_VC_delta_tdh_fractional_expression.tsv', norm = 'fractional',\n",
    "               name = 'Sprenger_2024_VC_delta_tdh_VP882_WT', year = 2024, journal = 'CellPress', firstauthor = 'Sprenger', pubmedID = 38579715, description = description, host_id=10, phage_id=7, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Sprenger_2024/Notebook_2/Sprenger_VC_delta_tdh_full_TPM.tsv', norm = 'TPM',\n",
    "               name = 'Sprenger_2024_VC_delta_tdh_VP882_WT', year = 2024, journal = 'CellPress', firstauthor = 'Sprenger', pubmedID = 38579715, description = description, host_id=10, phage_id=7, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Sprenger_2024/Notebook_2/Sprenger_VC_delta_tdh_TPM_means.tsv', norm = 'TPM_means',\n",
    "               name = 'Sprenger_2024_VC_delta_tdh_VP882_WT', year = 2024, journal = 'CellPress', firstauthor = 'Sprenger', pubmedID = 38579715, description = description, host_id=10, phage_id=7, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Sprenger_2024/Notebook_2/Sprenger_VC_delta_tdh_TPM_std.tsv', norm = 'TPM_std',\n",
    "               name = 'Sprenger_2024_VC_delta_tdh_VP882_WT', year = 2024, journal = 'CellPress', firstauthor = 'Sprenger', pubmedID = 38579715, description = description, host_id=10, phage_id=7, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Sprenger_2024/Notebook_2/Sprenger_VC_delta_tdh_full_raw_counts.tsv', norm = 'raw_counts',\n",
    "               name = 'Sprenger_2024_VC_delta_tdh_VP882_WT', year = 2024, journal = 'CellPress', firstauthor = 'Sprenger', pubmedID = 38579715, description = description, host_id=10, phage_id=7, doi = doi, upload_date = '25-08-12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "id": "073f372c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [15] Add datasets meaden_2021 P. aeruginosa PA14 WT infection\n",
    "\n",
    "description = 'Infection of P. aeruginosa PA14 WT with Pseudomonas phage DMS3 (MOI = 0.5). Samples were taken 35, 60 and 120 minutes post-infection (n=5). [PRJEB31514]'\n",
    "doi = 'https://doi.org/10.1038/s41396-020-00794-w'\n",
    "\n",
    "with app.app_context():\n",
    "    addFullDataset(phageName='DMS3', phageDes = 'Pseudomonas phage DMS3', phageID = 'NC_008717.1', phageType = 'temperate',\n",
    "               hostName = 'P. aeruginosa PA14 WT', hostDes = 'Pseudomonas aeruginosa PA14 WT', hostID = 'NZ_CP136842.1', hostGroup = 'Pseudomonas aeruginosa',\n",
    "               rnaSeqPath = 'Datasets/Meaden_2021/Notebook/Meaden_WT_fractional_expression.tsv', norm = 'fractional',\n",
    "               name = 'Meaden_2021_WT_infection', year = 2021, journal = 'ISME', firstauthor = 'Meaden', pubmedID = 33011743, description = description, doi = doi, upload_date = '25-08-12',\n",
    "               phageGenomePath = 'Datasets/Meaden_2021/Notebook/Meaden_WT_phage_gff3.tsv', phageGenomeName = 'Pseudomonas phage DMS3 genome',\n",
    "               hostGenomePath = 'Datasets/Meaden_2021/Notebook/Meaden_WT_host_gff3.tsv', hostGenomeName = 'P. aeruginosa PA14 WT genome')\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Meaden_2021/Notebook/Meaden_WT_full_TPM.tsv', norm = 'TPM',\n",
    "               name = 'Meaden_2021_WT_infection', year = 2021, journal = 'ISME', firstauthor = 'Meaden', pubmedID = 33011743, description = description, host_id=11, phage_id=9, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Meaden_2021/Notebook/Meaden_WT_TPM_means.tsv', norm = 'TPM_means',\n",
    "               name = 'Meaden_2021_WT_infection', year = 2021, journal = 'ISME', firstauthor = 'Meaden', pubmedID = 33011743, description = description, host_id=11, phage_id=9, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Meaden_2021/Notebook/Meaden_WT_TPM_std.tsv', norm = 'TPM_std',\n",
    "               name = 'Meaden_2021_WT_infection', year = 2021, journal = 'ISME', firstauthor = 'Meaden', pubmedID = 33011743, description = description, host_id=11, phage_id=9, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Meaden_2021/Notebook/Meaden_WT_full_raw_counts.tsv', norm = 'raw_counts',\n",
    "               name = 'Meaden_2021_WT_infection', year = 2021, journal = 'ISME', firstauthor = 'Meaden', pubmedID = 33011743, description = description, host_id=11, phage_id=9, doi = doi, upload_date = '25-08-12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "id": "f44456bf",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [16] Add datasets meaden_2021 P. aeruginosa PA14 BIM infection\n",
    "\n",
    "description = 'Infection of P. aeruginosa PA14 BIM2 (carrying two crispr spacers targeting DMS3vir) with Pseudomonas phage DMS3 (MOI = 0.5). Samples were taken 35, 60 and 120 minutes post-infection (n=5). [PRJEB31514]'\n",
    "doi = 'https://doi.org/10.1038/s41396-020-00794-w'\n",
    "\n",
    "with app.app_context():\n",
    "    hostAdd = Host(name = 'P. aeruginosa PA14 BIM2', description = 'Pseudomonas aeruginosa PA14 BIM2', ncbi_id = 'NZ_CP136842.1', group = 'Pseudomonas aeruginosa')\n",
    "    db.session.add(hostAdd)\n",
    "    db.session.commit()\n",
    "\n",
    "with app.app_context():\n",
    "    hostgff = pd.read_csv('Datasets/Meaden_2021/Notebook/Meaden_BIM_host_gff3.tsv', sep='\\t', comment = '#')\n",
    "    hostgenome = pickle.dumps(hostgff)\n",
    "    hostGFF = HostGenome(name = 'P. aeruginosa PA14 BIM2 genome', host_id = 12, gff_data = hostgenome)\n",
    "    db.session.add(hostGFF)\n",
    "    db.session.commit()\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Meaden_2021/Notebook/Meaden_BIM_fractional_expression.tsv', norm = 'fractional',\n",
    "               name = 'Meaden_2021_BIM2_infection', year = 2021, journal = 'ISME', firstauthor = 'Meaden', pubmedID = 33011743, description = description, host_id=12, phage_id=9, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Meaden_2021/Notebook/Meaden_BIM_full_TPM.tsv', norm = 'TPM',\n",
    "               name = 'Meaden_2021_BIM2_infection', year = 2021, journal = 'ISME', firstauthor = 'Meaden', pubmedID = 33011743, description = description, host_id=12, phage_id=9, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Meaden_2021/Notebook/Meaden_BIM_TPM_means.tsv', norm = 'TPM_means',\n",
    "               name = 'Meaden_2021_BIM2_infection', year = 2021, journal = 'ISME', firstauthor = 'Meaden', pubmedID = 33011743, description = description, host_id=12, phage_id=9, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Meaden_2021/Notebook/Meaden_BIM_TPM_std.tsv', norm = 'TPM_std',\n",
    "               name = 'Meaden_2021_BIM2_infection', year = 2021, journal = 'ISME', firstauthor = 'Meaden', pubmedID = 33011743, description = description, host_id=12, phage_id=9, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Meaden_2021/Notebook/Meaden_BIM_full_raw_counts.tsv', norm = 'raw_counts',\n",
    "               name = 'Meaden_2021_BIM2_infection', year = 2021, journal = 'ISME', firstauthor = 'Meaden', pubmedID = 33011743, description = description, host_id=12, phage_id=9, doi = doi, upload_date = '25-08-12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "id": "4077b675",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [17] Add datasets meaden_2021 P. aeruginosa PA14 BIM control\n",
    "\n",
    "description = 'Uninfected control of P. aeruginosa PA14 BIM2 (carrying two crispr spacers targeting DMS3vir) without addition of phage. Samples were taken 35, 60 and 120 minutes simultaneously to infected sample (n=5). [PRJEB31514]'\n",
    "doi = 'https://doi.org/10.1038/s41396-020-00794-w'\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Meaden_2021/Notebook/Meaden_BIM_Ctrl_fractional_expression.tsv', norm = 'fractional',\n",
    "               name = 'Meaden_2021_BIM2_control', year = 2021, journal = 'ISME', firstauthor = 'Meaden', pubmedID = 33011743, description = description, host_id=12, phage_id=9, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Meaden_2021/Notebook/Meaden_BIM_Ctrl_full_TPM.tsv', norm = 'TPM',\n",
    "               name = 'Meaden_2021_BIM2_control', year = 2021, journal = 'ISME', firstauthor = 'Meaden', pubmedID = 33011743, description = description, host_id=12, phage_id=9, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Meaden_2021/Notebook/Meaden_BIM_Ctrl_TPM_means.tsv', norm = 'TPM_means',\n",
    "               name = 'Meaden_2021_BIM2_control', year = 2021, journal = 'ISME', firstauthor = 'Meaden', pubmedID = 33011743, description = description, host_id=12, phage_id=9, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Meaden_2021/Notebook/Meaden_BIM_Ctrl_TPM_std.tsv', norm = 'TPM_std',\n",
    "               name = 'Meaden_2021_BIM2_control', year = 2021, journal = 'ISME', firstauthor = 'Meaden', pubmedID = 33011743, description = description, host_id=12, phage_id=9, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Meaden_2021/Notebook/Meaden_BIM_Ctrl_full_raw_counts.tsv', norm = 'raw_counts',\n",
    "               name = 'Meaden_2021_BIM2_control', year = 2021, journal = 'ISME', firstauthor = 'Meaden', pubmedID = 33011743, description = description, host_id=12, phage_id=9, doi = doi, upload_date = '25-08-12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "id": "dd87c225",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [18] Add Zhong_2020 datasets \n",
    "\n",
    "description = 'Phage-resistant mutant Pseudomonas aeruginosa strain PAO1(r_8) infected with dsRNA Bacteriophage phiYY (MOI = 10). Samples were taken before (0) and 6, 12, 18 min post-infection in biological triplicates (n=3). [GSE128811]'\n",
    "doi = 'https://doi.org/10.1016/j.isci.2020.101437'\n",
    "\n",
    "with app.app_context():\n",
    "    addFullDataset(phageName='phiYY', phageDes = 'Pseudomonas phage phiYY', phageID = 'KX074201.1 + KX074202.1 + KX074203.1', phageType = 'virulent',\n",
    "               hostName = 'P. aeruginosa PAO1r_8', hostDes = 'Pseudomonas aeruginosa PAO1r_8', hostID = 'NC_002516.2', hostGroup = 'Pseudomonas aeruginosa',\n",
    "               rnaSeqPath = 'Datasets/Zhong_2020/Notebook/Zhong_fractional_expression.tsv', norm = 'fractional',\n",
    "               name = 'Zhong_2020', year = 2020, journal = 'iScience', firstauthor = 'Zhong', pubmedID = 32827855, description = description, doi = doi, upload_date = '25-08-12',\n",
    "               phageGenomePath = 'Datasets/Zhong_2020/Notebook/Zhong_phage_gff3.tsv', phageGenomeName = 'Pseudomonas phage phiYY genome',\n",
    "               hostGenomePath = 'Datasets/Zhong_2020/Notebook/Zhong_host_gff3.tsv', hostGenomeName = 'P. aeruginosa PAO1r_8 genome')\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Zhong_2020/Notebook/Zhong_full_TPM.tsv', norm = 'TPM',\n",
    "               name = 'Zhong_2020', year = 2020, journal = 'iScience', firstauthor = 'Zhong', pubmedID = 32827855, description = description, host_id=13, phage_id=10, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Zhong_2020/Notebook/Zhong_TPM_means.tsv', norm = 'TPM_means',\n",
    "               name = 'Zhong_2020', year = 2020, journal = 'iScience', firstauthor = 'Zhong', pubmedID = 32827855, description = description, host_id=13, phage_id=10, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Zhong_2020/Notebook/Zhong_TPM_std.tsv', norm = 'TPM_std',\n",
    "               name = 'Zhong_2020', year = 2020, journal = 'iScience', firstauthor = 'Zhong', pubmedID = 32827855, description = description, host_id=13, phage_id=10, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Zhong_2020/Notebook/Zhong_full_raw_counts.tsv', norm = 'raw_counts',\n",
    "               name = 'Zhong_2020', year = 2020, journal = 'iScience', firstauthor = 'Zhong', pubmedID = 32827855, description = description, host_id=13, phage_id=10, doi = doi, upload_date = '25-08-12')\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "id": "c5b2aa47",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [19] Add Lood_2020 datasets\n",
    "\n",
    "description = 'Pseudomonas aeruginosa strain PAO1 infected with PA5oct (MOI = 50). Samples were taken before infection at 0 min (n=4) and 5, 15, 25 min post-infection (n=3). [GSE130190]'\n",
    "doi = 'https://doi.org/10.1111/1462-2920.14979'\n",
    "\n",
    "with app.app_context():\n",
    "    phageAdd = Phage(name = 'vB_PaeM_PA5oct', description = 'Pseudomonas phage vB_PaeM_PA5oct', ncbi_id = 'MK797984.1', phage_type = 'virulent')\n",
    "    db.session.add(phageAdd)\n",
    "    db.session.commit()\n",
    "\n",
    "with app.app_context():\n",
    "    phagegff = pd.read_csv('Datasets/Lood_2020/Notebook/Lood_phage_gff3.tsv', sep='\\t', comment = '#')\n",
    "    phagegenome = pickle.dumps(phagegff)\n",
    "    phageGFF = PhageGenome(name = 'vB_PaeM_PA5oct genome', phage_id = 11, gff_data = phagegenome)\n",
    "    db.session.add(phageGFF)\n",
    "    db.session.commit()\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Lood_2020/Notebook/Lood_fractional_expression.tsv', norm = 'fractional',\n",
    "               name = 'Lood_2020', year = 2020, journal = 'Environmental Microbiology', firstauthor = 'Lood', pubmedID = 32154616, description = description, host_id=8, phage_id=11, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Lood_2020/Notebook/Lood_full_TPM.tsv', norm = 'TPM',\n",
    "               name = 'Lood_2020', year = 2020, journal = 'Environmental Microbiology', firstauthor = 'Lood', pubmedID = 32154616, description = description, host_id=8, phage_id=11, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Lood_2020/Notebook/Lood_TPM_means.tsv', norm = 'TPM_means',\n",
    "               name = 'Lood_2020', year = 2020, journal = 'Environmental Microbiology', firstauthor = 'Lood', pubmedID = 32154616, description = description, host_id=8, phage_id=11, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Lood_2020/Notebook/Lood_TPM_std.tsv', norm = 'TPM_std',\n",
    "               name = 'Lood_2020', year = 2020, journal = 'Environmental Microbiology', firstauthor = 'Lood', pubmedID = 32154616, description = description, host_id=8, phage_id=11, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Lood_2020/Notebook/Lood_full_raw_counts.tsv', norm = 'raw_counts',\n",
    "               name = 'Lood_2020', year = 2020, journal = 'Environmental Microbiology', firstauthor = 'Lood', pubmedID = 32154616, description = description, host_id=8, phage_id=11, doi = doi, upload_date = '25-08-12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "id": "b38ec8f8",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [20] Add Li_2020 datasets\n",
    "\n",
    "description = 'Infection of Clostridioides difficile Ribotype 078 with phage JD032 (MOI = 10). Samples were taken before (0) and 30, 45, 75, 135 min post-infection in biological triplicates (n=3). [PRJNA559590]'\n",
    "doi = 'https://doi.org/10.1128/mSystems.00017-20'\n",
    "\n",
    "with app.app_context():\n",
    "    addFullDataset(phageName='JD032', phageDes = 'Clostridioides phage JD032', phageID = 'MK473382.1', phageType = 'virulent',\n",
    "               hostName = 'C. difficile RT078', hostDes = 'Clostridiodes difficile RT078 strain TW11', hostID = 'CP045224.1 + CP045225.1', hostGroup = 'Clostridiodes difficile',\n",
    "               rnaSeqPath = 'Datasets/Li_2020/Notebook/Li_fractional_expression.tsv', norm = 'fractional',\n",
    "               name = 'Li_2020', year = 2020, journal = 'mSystems', firstauthor = 'Li', pubmedID = 32371470, description = description, doi = doi, upload_date = '25-08-12',\n",
    "               phageGenomePath = 'Datasets/Li_2020/Notebook/Li_phage_gff3.tsv', phageGenomeName = 'Clostridioides phage JD032 genome',\n",
    "               hostGenomePath = 'Datasets/Li_2020/Notebook/Li_host_gff3.tsv', hostGenomeName = 'C. difficile RT078 genome')\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Li_2020/Notebook/Li_full_TPM.tsv', norm = 'TPM',\n",
    "               name = 'Li_2020', year = 2020, journal = 'mSystems', firstauthor = 'Li', pubmedID = 32371470, description = description, host_id=14, phage_id=12, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Li_2020/Notebook/Li_TPM_means.tsv', norm = 'TPM_means',\n",
    "               name = 'Li_2020', year = 2020, journal = 'mSystems', firstauthor = 'Li', pubmedID = 32371470, description = description, host_id=14, phage_id=12, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Li_2020/Notebook/Li_TPM_std.tsv', norm = 'TPM_std',\n",
    "               name = 'Li_2020', year = 2020, journal = 'mSystems', firstauthor = 'Li', pubmedID = 32371470, description = description, host_id=14, phage_id=12, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Li_2020/Notebook/Li_full_raw_counts.tsv', norm = 'raw_counts',\n",
    "               name = 'Li_2020', year = 2020, journal = 'mSystems', firstauthor = 'Li', pubmedID = 32371470, description = description, host_id=14, phage_id=12, doi = doi, upload_date = '25-08-12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "id": "1b438258",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [21] Add Yang_2019 datasets Acinetobacter baumannii AB1 infection\n",
    "\n",
    "description = 'Infection of drug resistant Acinetobacter baumannii AB1 isolated from hospital patient with phage phiAbp1 (MOI = 10). Samples were taken 5, 15, 30 min post-infection in biological triplicates (n=3). [GSE117396]'\n",
    "doi = 'https://doi.org/10.1128/msystems.00068-19'\n",
    "\n",
    "with app.app_context():\n",
    "    addFullDataset(phageName='phiAbp1', phageDes = 'Acinetobacter phage phiAbp1', phageID = 'NC_021316.1', phageType = 'virulent',\n",
    "               hostName = 'A. baumannii AB1', hostDes = 'Acinetobacter baumannii AB1 strain ATCC 17978-VU', hostID = 'NZ_CP018664.1', hostGroup = 'Acinetobacter baumannii',\n",
    "               rnaSeqPath = 'Datasets/Yang_2019/Notebook/Yang_fractional_expression.tsv', norm = 'fractional',\n",
    "               name = 'Yang_2019_infection', year = 2019, journal = 'mSystems', firstauthor = 'Yang', pubmedID = 31020041, description = description, doi = doi, upload_date = '25-08-12',\n",
    "               phageGenomePath = 'Datasets/Yang_2019/Notebook/Yang_phage_gff3.tsv', phageGenomeName = 'Acinetobacter phage Abp1 genome',\n",
    "               hostGenomePath = 'Datasets/Yang_2019/Notebook/Yang_host_gff3.tsv', hostGenomeName = 'A. baumannii AB1 genome')\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Yang_2019/Notebook/Yang_full_TPM.tsv', norm = 'TPM',\n",
    "               name = 'Yang_2019_infection', year = 2019, journal = 'mSystems', firstauthor = 'Yang', pubmedID = 31020041, description = description, host_id=15, phage_id=13, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Yang_2019/Notebook/Yang_TPM_means.tsv', norm = 'TPM_means',\n",
    "               name = 'Yang_2019_infection', year = 2019, journal = 'mSystems', firstauthor = 'Yang', pubmedID = 31020041, description = description, host_id=15, phage_id=13, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Yang_2019/Notebook/Yang_TPM_std.tsv', norm = 'TPM_std',\n",
    "               name = 'Yang_2019_infection', year = 2019, journal = 'mSystems', firstauthor = 'Yang', pubmedID = 31020041, description = description, host_id=15, phage_id=13, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Yang_2019/Notebook/Yang_full_raw_counts.tsv', norm = 'raw_counts',\n",
    "               name = 'Yang_2019_infection', year = 2019, journal = 'mSystems', firstauthor = 'Yang', pubmedID = 31020041, description = description, host_id=15, phage_id=13, doi = doi, upload_date = '25-08-12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "id": "113e0d64",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [22] Add Yang_2019 datasets Acinetobacter baumannii AB1 uninfected control\n",
    "\n",
    "description = 'Uninfected control of drug resistant Acinetobacter baumannii AB1 isolated from hospital patient without addition of phage. Samples were collected at 5, 15, 30 min simultaneously to infected sample in biological triplicates (n=3). [GSE117396]'\n",
    "doi = 'https://doi.org/10.1128/msystems.00068-19'\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Yang_2019/Notebook/Yang_ctrl_fractional_expression.tsv', norm = 'fractional',\n",
    "               name = 'Yang_2019_control', year = 2019, journal = 'mSystems', firstauthor = 'Yang', pubmedID = 31020041, description = description, host_id=15, phage_id=13, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Yang_2019/Notebook/Yang_ctrl_full_TPM.tsv', norm = 'TPM',\n",
    "               name = 'Yang_2019_control', year = 2019, journal = 'mSystems', firstauthor = 'Yang', pubmedID = 31020041, description = description, host_id=15, phage_id=13, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Yang_2019/Notebook/Yang_ctrl_TPM_means.tsv', norm = 'TPM_means',\n",
    "               name = 'Yang_2019_control', year = 2019, journal = 'mSystems', firstauthor = 'Yang', pubmedID = 31020041, description = description, host_id=15, phage_id=13, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Yang_2019/Notebook/Yang_ctrl_TPM_std.tsv', norm = 'TPM_std',\n",
    "               name = 'Yang_2019_control', year = 2019, journal = 'mSystems', firstauthor = 'Yang', pubmedID = 31020041, description = description, host_id=15, phage_id=13, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Yang_2019/Notebook/Yang_ctrl_full_raw_counts.tsv', norm = 'raw_counts',\n",
    "               name = 'Yang_2019_control', year = 2019, journal = 'mSystems', firstauthor = 'Yang', pubmedID = 31020041, description = description, host_id=15, phage_id=13, doi = doi, upload_date = '25-08-12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "id": "8a33ac0a",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [23] Add Ceyssens_2014 datasets (directional sequencing)\n",
    "\n",
    "description = 'Infection of Pseudomonas aeruginosa PAO1 with bacteriophage phiKZ (MOI = 5). Samples were taken before (0) and 5, 15, 35 min post-infection in biological duplicates (n=2). Samples were analyzed by directional RNA sequencing (5x10^6 reads/sample). [GSE58494]'\n",
    "doi = 'https://doi.org/10.1128/jvi.01347-14'\n",
    "\n",
    "with app.app_context():\n",
    "    phageAdd = Phage(name = 'phiKZ', description = 'Pseudomonas phage phiKZ', ncbi_id = 'NC_004629.1', phage_type = 'virulent')\n",
    "    db.session.add(phageAdd)\n",
    "    db.session.commit()\n",
    "\n",
    "with app.app_context():\n",
    "    phagegff = pd.read_csv('Datasets/Ceyssens_2014/Notebook/Ceyssens_directional_phage_gff3.tsv', sep='\\t', comment = '#')\n",
    "    phagegenome = pickle.dumps(phagegff)\n",
    "    phageGFF = PhageGenome(name = 'phiKZ genome', phage_id = 14, gff_data = phagegenome)\n",
    "    db.session.add(phageGFF)\n",
    "    db.session.commit()\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Ceyssens_2014/Notebook/Ceyssens_directional_fractional_expression.tsv', norm = 'fractional',\n",
    "               name = 'Ceyssens_2014_directional', year = 2014, journal = 'Journal of Virology', firstauthor = 'Ceyssens', pubmedID = 24965474, description = description, host_id=8, phage_id=14, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Ceyssens_2014/Notebook/Ceyssens_directional_full_TPM.tsv', norm = 'TPM',\n",
    "               name = 'Ceyssens_2014_directional', year = 2014, journal = 'Journal of Virology', firstauthor = 'Ceyssens', pubmedID = 24965474, description = description, host_id=8, phage_id=14, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Ceyssens_2014/Notebook/Ceyssens_directional_TPM_means.tsv', norm = 'TPM_means',\n",
    "               name = 'Ceyssens_2014_directional', year = 2014, journal = 'Journal of Virology', firstauthor = 'Ceyssens', pubmedID = 24965474, description = description, host_id=8, phage_id=14, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Ceyssens_2014/Notebook/Ceyssens_directional_TPM_std.tsv', norm = 'TPM_std',\n",
    "               name = 'Ceyssens_2014_directional', year = 2014, journal = 'Journal of Virology', firstauthor = 'Ceyssens', pubmedID = 24965474, description = description, host_id=8, phage_id=14, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Ceyssens_2014/Notebook/Ceyssens_directional_full_raw_counts.tsv', norm = 'raw_counts',\n",
    "               name = 'Ceyssens_2014_directional', year = 2014, journal = 'Journal of Virology', firstauthor = 'Ceyssens', pubmedID = 24965474, description = description, host_id=8, phage_id=14, doi = doi, upload_date = '25-08-12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "id": "9c6374a4",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [24] Add Ceyssens_2014 datasets (non-directional sequencing)\n",
    "\n",
    "description = 'Infection of Pseudomonas aeruginosa PAO1 with bacteriophage phiKZ (MOI = 5). Samples were taken 10 and 35 min post-infection in biological duplicates (n=2). Samples were analyzed by deeper (nondirectional) RNA sequencing (5x10e7 reads/sample). [GSE58494]'\n",
    "doi = 'https://doi.org/10.1128/jvi.01347-14'\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Ceyssens_2014/Notebook/Ceyssens_non-directional_fractional_expression.tsv', norm = 'fractional',\n",
    "               name = 'Ceyssens_2014_non-directional', year = 2014, journal = 'Journal of Virology', firstauthor = 'Ceyssens', pubmedID = 24965474, description = description, host_id=8, phage_id=14, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Ceyssens_2014/Notebook/Ceyssens_non-directional_full_TPM.tsv', norm = 'TPM',\n",
    "               name = 'Ceyssens_2014_non-directional', year = 2014, journal = 'Journal of Virology', firstauthor = 'Ceyssens', pubmedID = 24965474, description = description, host_id=8, phage_id=14, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Ceyssens_2014/Notebook/Ceyssens_non-directional_TPM_means.tsv', norm = 'TPM_means',\n",
    "               name = 'Ceyssens_2014_non-directional', year = 2014, journal = 'Journal of Virology', firstauthor = 'Ceyssens', pubmedID = 24965474, description = description, host_id=8, phage_id=14, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Ceyssens_2014/Notebook/Ceyssens_non-directional_TPM_std.tsv', norm = 'TPM_std',\n",
    "               name = 'Ceyssens_2014_non-directional', year = 2014, journal = 'Journal of Virology', firstauthor = 'Ceyssens', pubmedID = 24965474, description = description, host_id=8, phage_id=14, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Ceyssens_2014/Notebook/Ceyssens_non-directional_full_raw_counts.tsv', norm = 'raw_counts',\n",
    "               name = 'Ceyssens_2014_non-directional', year = 2014, journal = 'Journal of Virology', firstauthor = 'Ceyssens', pubmedID = 24965474, description = description, host_id=8, phage_id=14, doi = doi, upload_date = '25-08-12')"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "beb9cb22",
   "metadata": {},
   "source": [
    "# Datasets Round 2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "61308571",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [25] Add Kongari_2024 dataset (wt)\n",
    "\n",
    "description = 'Infection of clinical isolate Staphyloccocus aureus USA300 strain NRS384 with Staphylococcus phage K (MOI = 5). Samples were taken before (0) and 2, 5, 10, 20, 30, 40 min post infection (n=4). [GSE253516]'\n",
    "doi = 'https://doi.org/10.3390/v16111773'\n",
    "\n",
    "with app.app_context():\n",
    "    addFullDataset(phageName='Phage K', phageDes = 'Staphylococcus Phage K', phageID = 'NC_005880.2', phageType = 'virulent',\n",
    "               hostName = 'S. aureus USA300 NRS384', hostDes = 'Staphylococcus aureus USA300 clinical isolate strain NRS384', hostID = 'CP027476.1', hostGroup = 'Staphylococcus aureus',\n",
    "               rnaSeqPath = 'Datasets/Kongari_2024/notebook/Kongari_WT_fractional_expression.tsv', norm = 'fractional',\n",
    "               name = 'Kongari_2024_WT', year = 2024, journal = 'Viruses', firstauthor = 'Kongari', pubmedID = 39599887, description = description, doi = doi, upload_date = '25-08-12',\n",
    "               phageGenomePath = 'Datasets/Kongari_2024/notebook/Kongari_WT_phage_gff3.tsv', phageGenomeName = 'Staphylococcus phage K genome',\n",
    "               hostGenomePath = 'Datasets/Kongari_2024/notebook/Kongari_WT_host_gff3.tsv', hostGenomeName = 'S. aureus USA300 NRS384 genome')\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Kongari_2024/notebook/Kongari_WT_full_TPM.tsv', norm = 'TPM',\n",
    "               name = 'Kongari_2024_WT', year = 2024, journal = 'Viruses', firstauthor = 'Kongari', pubmedID = 39599887, description = description, host_id=16, phage_id=15, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Kongari_2024/notebook/Kongari_WT_TPM_means.tsv', norm = 'TPM_means',\n",
    "               name = 'Kongari_2024_WT', year = 2024, journal = 'Viruses', firstauthor = 'Kongari', pubmedID = 39599887, description = description, host_id=16, phage_id=15, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Kongari_2024/notebook/Kongari_WT_TPM_std.tsv', norm = 'TPM_std',\n",
    "               name = 'Kongari_2024_WT', year = 2024, journal = 'Viruses', firstauthor = 'Kongari', pubmedID = 39599887, description = description, host_id=16, phage_id=15, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Kongari_2024/notebook/Kongari_WT_full_raw_counts.tsv', norm = 'raw_counts',\n",
    "               name = 'Kongari_2024_WT', year = 2024, journal = 'Viruses', firstauthor = 'Kongari', pubmedID = 39599887, description = description, host_id=16, phage_id=15, doi = doi, upload_date = '25-08-12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "id": "d1ec3e0b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [26] Add Kongari_2024 dataset (rpoC G17D)\n",
    "\n",
    "description = 'Infection of clinical isolate Staphyloccocus aureses USA300 strain NRS384 rpoC G17D mutant with Staphylococcus phage K (MOI = 5). Samples were taken before (0) and 2, 5, 10, 20, 30, 40 min post infection (n=4). [GSE253516]'\n",
    "doi = 'https://doi.org/10.3390/v16111773'\n",
    "\n",
    "with app.app_context():\n",
    "    hostAdd = Host(name = 'S. aureus USA300 NRS384 rpoC G17D', description = 'Staphylococcus aureus USA300 clinical isolate strain NRS384 rpoC G17D', ncbi_id = 'CP027476.1', group = 'Staphylococcus aureus')\n",
    "    db.session.add(hostAdd)\n",
    "    db.session.commit()\n",
    "\n",
    "with app.app_context():\n",
    "    hostgff = pd.read_csv('Datasets/Kongari_2024/notebook/Kongari_rpoC_G17D_host_gff3.tsv', sep='\\t', comment = '#')\n",
    "    hostgenome = pickle.dumps(hostgff)\n",
    "    hostGFF = HostGenome(name = 'S. aureus USA300 NRS384 rpoC G17D genome', host_id = 17, gff_data = hostgenome)\n",
    "    db.session.add(hostGFF)\n",
    "    db.session.commit()\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Kongari_2024/notebook/Kongari_rpoC_G17D_fractional_expression.tsv', norm = 'fractional',\n",
    "               name = 'Kongari_2024_rpoC-G17D', year = 2024, journal = 'Viruses', firstauthor = 'Kongari', pubmedID = 39599887, description = description, host_id=17, phage_id=15, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Kongari_2024/notebook/Kongari_rpoC_G17D_full_TPM.tsv', norm = 'TPM',\n",
    "               name = 'Kongari_2024_rpoC-G17D', year = 2024, journal = 'Viruses', firstauthor = 'Kongari', pubmedID = 39599887, description = description, host_id=17, phage_id=15, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Kongari_2024/notebook/Kongari_rpoC_G17D_TPM_means.tsv', norm = 'TPM_means',\n",
    "               name = 'Kongari_2024_rpoC-G17D', year = 2024, journal = 'Viruses', firstauthor = 'Kongari', pubmedID = 39599887, description = description, host_id=17, phage_id=15, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Kongari_2024/notebook/Kongari_rpoC_G17D_TPM_std.tsv', norm = 'TPM_std',\n",
    "               name = 'Kongari_2024_rpoC-G17D', year = 2024, journal = 'Viruses', firstauthor = 'Kongari', pubmedID = 39599887, description = description, host_id=17, phage_id=15, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Kongari_2024/notebook/Kongari_rpoC_G17D_full_raw_counts.tsv', norm = 'raw_counts',\n",
    "               name = 'Kongari_2024_rpoC-G17D', year = 2024, journal = 'Viruses', firstauthor = 'Kongari', pubmedID = 39599887, description = description, host_id=17, phage_id=15, doi = doi, upload_date = '25-08-12')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "297b3549",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [27] Add Gerovac_1_2024 dataset\n",
    "\n",
    "description = 'Infection of exponentially growing Pseudomonas aeruginosa PAO1 in M9 (+Glucose) with Pseudomonas phage phiKZ (MOI = 15). Samples were taken before (0) and 2, 4, 6, 8, 10 min post infection in biological duplicates (n=2). Only early infection phase is captured, thus no classification of phage genes (early, middle, late) is performed. [GSE223979]'\n",
    "doi = 'https://doi.org/10.1038/s41564-024-01616-x'\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Gerovac_2024_1/notebook/Gerovac_fractional_expression.tsv', norm = 'fractional',\n",
    "               name = 'Gerovac_1_2024', year = 2024, journal = 'Nature microbiology', firstauthor = 'Gerovac', pubmedID = 38443577, description = description, host_id=8, phage_id=14, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Gerovac_2024_1/notebook/Gerovac_full_TPM.tsv', norm = 'TPM',\n",
    "               name = 'Gerovac_1_2024', year = 2024, journal = 'Nature microbiology', firstauthor = 'Gerovac', pubmedID = 38443577, description = description, host_id=8, phage_id=14, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Gerovac_2024_1/notebook/Gerovac_TPM_means.tsv', norm = 'TPM_means',\n",
    "               name = 'Gerovac_1_2024', year = 2024, journal = 'Nature microbiology', firstauthor = 'Gerovac', pubmedID = 38443577, description = description, host_id=8, phage_id=14, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Gerovac_2024_1/notebook/Gerovac_TPM_std.tsv', norm = 'TPM_std',\n",
    "               name = 'Gerovac_1_2024', year = 2024, journal = 'Nature microbiology', firstauthor = 'Gerovac', pubmedID = 38443577, description = description, host_id=8, phage_id=14, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Gerovac_2024_1/notebook/Gerovac_full_raw_counts.tsv', norm = 'raw_counts',\n",
    "               name = 'Gerovac_1_2024', year = 2024, journal = 'Nature microbiology', firstauthor = 'Gerovac', pubmedID = 38443577, description = description, host_id=8, phage_id=14, doi = doi, upload_date = '25-08-12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "271a043b",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [28] Add Gerovac_2_2024_ChmA-KD dataset\n",
    "\n",
    "description = 'Infection of Pseudomonas aeruginosa PAO1 ChmA knockdown (pretreatment with chmA transcript ASO targeting ribosomal binding site and start codon) with Pseudomonas phage phiKZ (MOI = 5). Samples were taken 10, 15, 20, 25, 35 min post infection in biological duplicates (n=2). [GSE269401]'\n",
    "doi = 'https://doi.org/10.1101/2024.07.31.605949'\n",
    "\n",
    "with app.app_context():\n",
    "    phageAdd = Phage(name = 'phiKZ ChmA-KD', description = 'Pseudomonas phage phiKZ ChmA-KD', ncbi_id = 'NC_004629.1', phage_type = 'virulent')\n",
    "    db.session.add(phageAdd)\n",
    "    db.session.commit()\n",
    "\n",
    "with app.app_context():\n",
    "    phagegff = pd.read_csv('Datasets/Gerovac_2024_2/notebook/Gerovac_ChmA-KD_phage_gff3.tsv', sep='\\t', comment = '#')\n",
    "    phagegenome = pickle.dumps(phagegff)\n",
    "    phageGFF = PhageGenome(name = 'phiKZ ChmA-KD genome', phage_id = 16, gff_data = phagegenome)\n",
    "    db.session.add(phageGFF)\n",
    "    db.session.commit()\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Gerovac_2024_2/notebook/Gerovac_ChmA-KD_fractional_expression.tsv', norm = 'fractional',\n",
    "               name = 'Gerovac_2_2024', year = 2024, journal = 'BioRxiv', firstauthor = 'Gerovac', pubmedID = \"preprint\", description = description, host_id=8, phage_id=16, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Gerovac_2024_2/notebook/Gerovac_ChmA-KD_full_TPM.tsv', norm = 'TPM',\n",
    "               name = 'Gerovac_2_2024', year = 2024, journal = 'BioRxiv', firstauthor = 'Gerovac', pubmedID = \"preprint\", description = description, host_id=8, phage_id=16, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Gerovac_2024_2/notebook/Gerovac_ChmA-KD_TPM_means.tsv', norm = 'TPM_means',\n",
    "               name = 'Gerovac_2_2024', year = 2024, journal = 'BioRxiv', firstauthor = 'Gerovac', pubmedID = \"preprint\", description = description, host_id=8, phage_id=16, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Gerovac_2024_2/notebook/Gerovac_ChmA-KD_TPM_std.tsv', norm = 'TPM_std',\n",
    "               name = 'Gerovac_2_2024', year = 2024, journal = 'BioRxiv', firstauthor = 'Gerovac', pubmedID = \"preprint\", description = description, host_id=8, phage_id=16, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Gerovac_2024_2/notebook/Gerovac_ChmA-KD_full_raw_counts.tsv', norm = 'raw_counts',\n",
    "               name = 'Gerovac_2_2024', year = 2024, journal = 'BioRxiv', firstauthor = 'Gerovac', pubmedID = \"preprint\", description = description, host_id=8, phage_id=16, doi = doi, upload_date = '25-08-12')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "id": "43085b05",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [29] Add Gerovac_2_2024_WT dataset\n",
    "\n",
    "description = 'Infection of Pseudomonas aeruginosa PAO1 with Pseudomonas phage phiKZ (MOI = 5). Samples were taken 10, 15, 20, 25, 35 min post infection in biological duplicates (n=2). [GSE269401]'\n",
    "doi = 'https://doi.org/10.1101/2024.07.31.605949'\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Gerovac_2024_2/notebook/Gerovac_WT_fractional_expression.tsv', norm = 'fractional',\n",
    "               name = 'Gerovac_2_2024', year = 2024, journal = 'BioRxiv', firstauthor = 'Gerovac', pubmedID = \"preprint\", description = description, host_id=8, phage_id=14, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Gerovac_2024_2/notebook/Gerovac_WT_full_TPM.tsv', norm = 'TPM',\n",
    "               name = 'Gerovac_2_2024', year = 2024, journal = 'BioRxiv', firstauthor = 'Gerovac', pubmedID = \"preprint\", description = description, host_id=8, phage_id=14, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Gerovac_2024_2/notebook/Gerovac_WT_TPM_means.tsv', norm = 'TPM_means',\n",
    "               name = 'Gerovac_2_2024', year = 2024, journal = 'BioRxiv', firstauthor = 'Gerovac', pubmedID = \"preprint\", description = description, host_id=8, phage_id=14, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Gerovac_2024_2/notebook/Gerovac_WT_TPM_std.tsv', norm = 'TPM_std',\n",
    "               name = 'Gerovac_2_2024', year = 2024, journal = 'BioRxiv', firstauthor = 'Gerovac', pubmedID = \"preprint\", description = description, host_id=8, phage_id=14, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Gerovac_2024_2/notebook/Gerovac_WT_full_raw_counts.tsv', norm = 'raw_counts',\n",
    "               name = 'Gerovac_2_2024', year = 2024, journal = 'BioRxiv', firstauthor = 'Gerovac', pubmedID = \"preprint\", description = description, host_id=8, phage_id=14, doi = doi, upload_date = '25-08-12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "f12cd4f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [30] Add Silva_exponential dataset\n",
    "\n",
    "description = 'Infection of exponential cultures of S. epidermidis 9142 with Staphylococcus phage SEP1 (MOI=5). Samples were taken before (0) and 5, 15, 30 min post infection in biological triplicates (n=3). [GSE254200]'\n",
    "doi = 'https://doi.org/10.1128/msystems.00263-24'\n",
    "\n",
    "with app.app_context():\n",
    "    addFullDataset(phageName='SEP1', phageDes = 'Staphylococcus phage SEP1', phageID = 'NC_041928.1', phageType = 'virulent',\n",
    "               hostName = 'S. epidermidis 9142', hostDes = 'Staphylococcus epidermidis 9142', hostID = 'CP000029.1', hostGroup = 'Staphylococcus epidermidis',\n",
    "               rnaSeqPath = 'Datasets/Silva_2024/notebook/Silva_exponential_fractional_expression.tsv', norm = 'fractional',\n",
    "               name = 'Silva_2024_exponential', year = 2024, journal = 'mSystems', firstauthor = 'Silva', pubmedID = 38904376, description = description, doi = doi, upload_date = '25-08-12',\n",
    "               phageGenomePath = 'Datasets/Silva_2024/notebook/Silva_exponential_phage_gff3.tsv', phageGenomeName = 'Staphylococcus phage SEP1 genome',\n",
    "               hostGenomePath = 'Datasets/Silva_2024/notebook/Silva_exponential_host_gff3.tsv', hostGenomeName = 'S. epidermidis 9142 genome')\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Silva_2024/notebook/Silva_exponential_full_TPM.tsv', norm = 'TPM',\n",
    "               name = 'Silva_2024_exponential', year = 2024, journal = 'mSystems', firstauthor = 'Silva', pubmedID = 38904376, description = description, host_id=18, phage_id=17, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Silva_2024/notebook/Silva_exponential_TPM_means.tsv', norm = 'TPM_means',\n",
    "               name = 'Silva_2024_exponential', year = 2024, journal = 'mSystems', firstauthor = 'Silva', pubmedID = 38904376, description = description, host_id=18, phage_id=17, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Silva_2024/notebook/Silva_exponential_TPM_std.tsv', norm = 'TPM_std',\n",
    "               name = 'Silva_2024_exponential', year = 2024, journal = 'mSystems', firstauthor = 'Silva', pubmedID = 38904376, description = description, host_id=18, phage_id=17, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Silva_2024/notebook/Silva_exponential_full_raw_counts.tsv', norm = 'raw_counts',\n",
    "               name = 'Silva_2024_exponential', year = 2024, journal = 'mSystems', firstauthor = 'Silva', pubmedID = 38904376, description = description, host_id=18, phage_id=17, doi = doi, upload_date = '25-08-12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "id": "f0fd2eab",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [31] Add Silva_stationary dataset\n",
    "\n",
    "description = 'Infection of stationary cultures of S. epidermidis 9142 with Staphylococcus phage SEP1 (MOI=5). Samples were taken before (0) and 5, 15, 30 min post infection in biological triplicates (n=3). [GSE254200]'\n",
    "doi = 'https://doi.org/10.1128/msystems.00263-24'\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Silva_2024/notebook/Silva_stationary_fractional_expression.tsv', norm = 'fractional',\n",
    "               name = 'Silva_2024_stationary', year = 2024, journal = 'mSystems', firstauthor = 'Silva', pubmedID = 38904376, description = description, host_id=18, phage_id=17, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Silva_2024/notebook/Silva_stationary_full_TPM.tsv', norm = 'TPM',\n",
    "               name = 'Silva_2024_stationary', year = 2024, journal = 'mSystems', firstauthor = 'Silva', pubmedID = 38904376, description = description, host_id=18, phage_id=17, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Silva_2024/notebook/Silva_stationary_TPM_means.tsv', norm = 'TPM_means',\n",
    "               name = 'Silva_2024_stationary', year = 2024, journal = 'mSystems', firstauthor = 'Silva', pubmedID = 38904376, description = description, host_id=18, phage_id=17, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Silva_2024/notebook/Silva_stationary_TPM_std.tsv', norm = 'TPM_std',\n",
    "               name = 'Silva_2024_stationary', year = 2024, journal = 'mSystems', firstauthor = 'Silva', pubmedID = 38904376, description = description, host_id=18, phage_id=17, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Silva_2024/notebook/Silva_stationary_full_raw_counts.tsv', norm = 'raw_counts',\n",
    "               name = 'Silva_2024_stationary', year = 2024, journal = 'mSystems', firstauthor = 'Silva', pubmedID = 38904376, description = description, host_id=18, phage_id=17, doi = doi, upload_date = '25-08-12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "id": "536ddb37",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [32] Add McLaughlin dataset\n",
    "\n",
    "description = 'Infection of Caulobacter Crescentus CB15 with Caulobacter phage phiCbK (MOI = 10). Samples were taken before (0) and 15, 30, 45, 60, 75, 90 min post infection in biological triplicates (n=3). [GSE241056]'\n",
    "doi = 'https://doi.org/10.1371/journal.pgen.1011048'\n",
    "\n",
    "with app.app_context():\n",
    "    addFullDataset(phageName='phiCbK', phageDes = 'Caulobacter phage phiCbK', phageID = 'JX100813.1', phageType = 'virulent',\n",
    "               hostName = 'C. crescentus CB15', hostDes = 'Caulobacter crescentus CB15', hostID = 'CP001340.1', hostGroup = 'Caulobacter crescentus',\n",
    "               rnaSeqPath = 'Datasets/McLaughlin_2023/notebook/McLaughlin_fractional_expression.tsv', norm = 'fractional',\n",
    "               name = 'McLaughlin_2023', year = 2023, journal = 'PLOS Genetics', firstauthor = 'McLaughlin', pubmedID = 37972151, description = description, doi = doi, upload_date = '25-08-12',\n",
    "               phageGenomePath = 'Datasets/McLaughlin_2023/notebook/McLaughlin_phage_gff3.tsv', phageGenomeName = 'Caulobacter phage phiCbK genome',\n",
    "               hostGenomePath = 'Datasets/McLaughlin_2023/notebook/McLaughlin_host_gff3.tsv', hostGenomeName = 'C. crescentus CB15 genome')\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/McLaughlin_2023/notebook/McLaughlin_full_TPM.tsv', norm = 'TPM',\n",
    "               name = 'McLaughlin_2023', year = 2023, journal = 'PLOS Genetics', firstauthor = 'McLaughlin', pubmedID = 37972151, description = description, host_id=19, phage_id=18, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/McLaughlin_2023/notebook/McLaughlin_TPM_means.tsv', norm = 'TPM_means',\n",
    "               name = 'McLaughlin_2023', year = 2023, journal = 'PLOS Genetics', firstauthor = 'McLaughlin', pubmedID = 37972151, description = description, host_id=19, phage_id=18, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/McLaughlin_2023/notebook/McLaughlin_TPM_std.tsv', norm = 'TPM_std',\n",
    "               name = 'McLaughlin_2023', year = 2023, journal = 'PLOS Genetics', firstauthor = 'McLaughlin', pubmedID = 37972151, description = description, host_id=19, phage_id=18, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/McLaughlin_2023/notebook/McLaughlin_full_raw_counts.tsv', norm = 'raw_counts',\n",
    "               name = 'McLaughlin_2023', year = 2023, journal = 'PLOS Genetics', firstauthor = 'McLaughlin', pubmedID = 37972151, description = description, host_id=19, phage_id=18, doi = doi, upload_date = '25-08-12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "id": "3191ef2c",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [33] Add Bürkle_mono-infection dataset\n",
    "\n",
    "description = 'Infection of Pseudomonas aeruginosa PAO1 with Pseudomonas phage JG024. Samples were taken 8 and 24 min post-infection in biological duplicates (n=2). Phage genes were not classified into infection phases due to limited temporal resolution of sampling. [GSE271537]'\n",
    "doi = 'https://doi.org/10.1093/ismejo/wraf065'\n",
    "\n",
    "with app.app_context():\n",
    "    phageAdd = Phage(name = 'JG024', description = 'Pseudomonas phage JG024', ncbi_id = 'NC_017674.1', phage_type = 'virulent')\n",
    "    db.session.add(phageAdd)\n",
    "    db.session.commit()\n",
    "\n",
    "with app.app_context():\n",
    "    phagegff = pd.read_csv('Datasets/Bürkle_2025/notebook/Bürkle_JG024_phage_gff3.tsv', sep='\\t', comment = '#')\n",
    "    phagegenome = pickle.dumps(phagegff)\n",
    "    phageGFF = PhageGenome(name = 'JG024 genome', phage_id = 19, gff_data = phagegenome)\n",
    "    db.session.add(phageGFF)\n",
    "    db.session.commit()\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Bürkle_2025/notebook/Bürkle_JG024_fractional_expression.tsv', norm = 'fractional',\n",
    "               name = 'Bürkle_2025_mono-infection', year = 2025, journal = 'ISME', firstauthor = 'Bürkle', pubmedID = 40188480, description = description, host_id=8, phage_id=19, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Bürkle_2025/notebook/Bürkle_JG024_full_TPM.tsv', norm = 'TPM',\n",
    "               name = 'Bürkle_2025_mono-infection', year = 2025, journal = 'ISME', firstauthor = 'Bürkle', pubmedID = 40188480, description = description, host_id=8, phage_id=19, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Bürkle_2025/notebook/Bürkle_JG024_TPM_means.tsv', norm = 'TPM_means',\n",
    "               name = 'Bürkle_2025_mono-infection', year = 2025, journal = 'ISME', firstauthor = 'Bürkle', pubmedID = 40188480, description = description, host_id=8, phage_id=19, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Bürkle_2025/notebook/Bürkle_JG024_TPM_std.tsv', norm = 'TPM_std',\n",
    "               name = 'Bürkle_2025_mono-infection', year = 2025, journal = 'ISME', firstauthor = 'Bürkle', pubmedID = 40188480, description = description, host_id=8, phage_id=19, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Bürkle_2025/notebook/Bürkle_JG024_full_raw_counts.tsv', norm = 'raw_counts',\n",
    "               name = 'Bürkle_2025_mono-infection', year = 2025, journal = 'ISME', firstauthor = 'Bürkle', pubmedID = 40188480, description = description, host_id=8, phage_id=19, doi = doi, upload_date = '25-08-12')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "id": "5d94bc44",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [34] Add Bürkle co-infection dataset\n",
    "\n",
    "description = 'Co-infection of Pseudomonas aeruginosa PAO1 with Pseudomonas phage JG024 and JG005. Samples were taken 8 and 24 min post-infection in biological duplicates (n=2). Phage genes were not classified into infection phases due to limited temporal resolution of sampling. [GSE271537]'\n",
    "doi = 'https://doi.org/10.1093/ismejo/wraf065'\n",
    "\n",
    "with app.app_context():\n",
    "    phageAdd = Phage(name = 'Phage cocktail (JG024 + JG005)', description = 'Phage cocktail (Pseudomonas phage JG024 + JG005)', ncbi_id = 'NC_017674.1 + PP712940.1', phage_type = 'virulent')\n",
    "    db.session.add(phageAdd)\n",
    "    db.session.commit()\n",
    "\n",
    "with app.app_context():\n",
    "    phagegff = pd.read_csv('Datasets/Bürkle_2025/notebook/Bürkle_phage-cocktail_phage_gff3.tsv', sep='\\t', comment = '#')\n",
    "    phagegenome = pickle.dumps(phagegff)\n",
    "    phageGFF = PhageGenome(name = 'Phage cocktail (JG024 + JG005) genomes', phage_id = 20, gff_data = phagegenome)\n",
    "    db.session.add(phageGFF)\n",
    "    db.session.commit()\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Bürkle_2025/notebook/Bürkle_phage-cocktail_fractional_expression.tsv', norm = 'fractional',\n",
    "               name = 'Bürkle_2025_co-infection', year = 2025, journal = 'ISME', firstauthor = 'Bürkle', pubmedID = 40188480, description = description, host_id=8, phage_id=20, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Bürkle_2025/notebook/Bürkle_phage-cocktail_full_TPM.tsv', norm = 'TPM',\n",
    "               name = 'Bürkle_2025_co-infection', year = 2025, journal = 'ISME', firstauthor = 'Bürkle', pubmedID = 40188480, description = description, host_id=8, phage_id=20, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Bürkle_2025/notebook/Bürkle_phage-cocktail_TPM_means.tsv', norm = 'TPM_means',\n",
    "               name = 'Bürkle_2025_co-infection', year = 2025, journal = 'ISME', firstauthor = 'Bürkle', pubmedID = 40188480, description = description, host_id=8, phage_id=20, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Bürkle_2025/notebook/Bürkle_phage-cocktail_TPM_std.tsv', norm = 'TPM_std',\n",
    "               name = 'Bürkle_2025_co-infection', year = 2025, journal = 'ISME', firstauthor = 'Bürkle', pubmedID = 40188480, description = description, host_id=8, phage_id=20, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Bürkle_2025/notebook/Bürkle_phage-cocktail_full_raw_counts.tsv', norm = 'raw_counts',\n",
    "               name = 'Bürkle_2025_co-infection', year = 2025, journal = 'ISME', firstauthor = 'Bürkle', pubmedID = 40188480, description = description, host_id=8, phage_id=20, doi = doi, upload_date = '25-08-12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "id": "46773f55",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [35] Add Buttimer dataset\n",
    " \n",
    "description = 'Infection of methicillin-resistent Staphylococcus aureus (MRSA) E1185(IV)ST12 with Staphyloccous phage SAM1 which was isolated from Fersisi commercial phage cocktail (MOI=10). Samples were taken 15, 35, 45 min post-infection in biological triplicates (n=3). [GSE192733]'\n",
    "doi = 'https://doi.org/10.3390/v14030626'\n",
    "\n",
    "with app.app_context():\n",
    "    addFullDataset(phageName='SAM1', phageDes = 'Staphylococcus phage SAM1', phageID = 'MT338525.1', phageType = 'virulent',\n",
    "               hostName = 'S. aureus E1185_IV_ST12', hostDes = 'Staphylococcus aureus strain E1185_IV_ST12', hostID = 'CP089586.1', hostGroup = 'Staphylococcus aureus',\n",
    "               rnaSeqPath = 'Datasets/Buttimer_2022/notebook/Buttimer_fractional_expression.tsv', norm = 'fractional',\n",
    "               name = 'Buttimer_2022', year = 2022, journal = 'Viruses', firstauthor = 'Buttimer', pubmedID = 35337034, description = description, doi = doi, upload_date = '25-08-12',\n",
    "               phageGenomePath = 'Datasets/Buttimer_2022/notebook/Buttimer_phage_gff3.tsv', phageGenomeName = 'Staphylococcus phage SAM1 genome',\n",
    "               hostGenomePath = 'Datasets/Buttimer_2022/notebook/Buttimer_host_gff3.tsv', hostGenomeName = 'S. aureus E1185_IV_ST12 genome')\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Buttimer_2022/notebook/Buttimer_full_TPM.tsv', norm = 'TPM',\n",
    "               name = 'Buttimer_2022', year = 2022, journal = 'Viruses', firstauthor = 'Buttimer', pubmedID = 35337034, description = description, host_id=20, phage_id=21, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Buttimer_2022/notebook/Buttimer_TPM_means.tsv', norm = 'TPM_means',\n",
    "               name = 'Buttimer_2022', year = 2022, journal = 'Viruses', firstauthor = 'Buttimer', pubmedID = 35337034, description = description, host_id=20, phage_id=21, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Buttimer_2022/notebook/Buttimer_TPM_std.tsv', norm = 'TPM_std',\n",
    "               name = 'Buttimer_2022', year = 2022, journal = 'Viruses', firstauthor = 'Buttimer', pubmedID = 35337034, description = description, host_id=20, phage_id=21, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Buttimer_2022/notebook/Buttimer_full_raw_counts.tsv', norm = 'raw_counts',\n",
    "               name = 'Buttimer_2022', year = 2022, journal = 'Viruses', firstauthor = 'Buttimer', pubmedID = 35337034, description = description, host_id=20, phage_id=21, doi = doi, upload_date = '25-08-12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "id": "dbedf9aa",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [36] Add Huang control dataset\n",
    "\n",
    "description = 'Uninfected control of picocyanobacteria Synechococcus strain WH7803 without addition of phage. Samples were collected simultaneously to infected samples at 15min, 1h, 3h, 5h and 7h (n=1). [GSE150732]'\n",
    "doi = 'https://doi.org/10.1002/mbo3.1150'\n",
    "\n",
    "with app.app_context():\n",
    "    addFullDataset(phageName='S-SBP1', phageDes = 'Synechococcus phage S-SBP1', phageID = 'MT424636.1', phageType = 'virulent',\n",
    "               hostName = 'Synechococcus WH7803', hostDes = 'Synechococcus sp. WH7803', hostID = 'CT971583.1', hostGroup = 'Synechococcus species',\n",
    "               rnaSeqPath = 'Datasets/Huang_2021/notebook/Huang_ctrl_fractional_expression.tsv', norm = 'fractional',\n",
    "               name = 'Haung_2021_control', year = 2021, journal = 'Microbiology Open', firstauthor = 'Huang', pubmedID = 33377630, description = description, doi = doi, upload_date = '25-08-12',\n",
    "               phageGenomePath = 'Datasets/Huang_2021/notebook/Huang_ctrl_phage_gff3.tsv', phageGenomeName = 'Synechococcus phage S-SBP1 genome',\n",
    "               hostGenomePath = 'Datasets/Huang_2021/notebook/Huang_ctrl_host_gff3.tsv', hostGenomeName = 'Synechococcus sp. WH7803 genome')\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Huang_2021/notebook/Huang_ctrl_full_TPM.tsv', norm = 'TPM',\n",
    "               name = 'Haung_2021_control', year = 2021, journal = 'Microbiology Open', firstauthor = 'Huang', pubmedID = 33377630, description = description, host_id=21, phage_id=22, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Huang_2021/notebook/Huang_ctrl_TPM_means.tsv', norm = 'TPM_means',\n",
    "               name = 'Haung_2021_control', year = 2021, journal = 'Microbiology Open', firstauthor = 'Huang', pubmedID = 33377630, description = description, host_id=21, phage_id=22, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Huang_2021/notebook/Huang_ctrl_TPM_std.tsv', norm = 'TPM_std',\n",
    "               name = 'Haung_2021_control', year = 2021, journal = 'Microbiology Open', firstauthor = 'Huang', pubmedID = 33377630, description = description, host_id=21, phage_id=22, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Huang_2021/notebook/Huang_ctrl_full_raw_counts.tsv', norm = 'raw_counts',\n",
    "               name = 'Haung_2021_control', year = 2021, journal = 'Microbiology Open', firstauthor = 'Huang', pubmedID = 33377630, description = description, host_id=21, phage_id=22, doi = doi, upload_date = '25-08-12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "id": "d7761298",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [37] Add Huang infection dataset\n",
    "\n",
    "description = 'Infection of picocyanobacteria Synechococcus strain WH7803 with Synechococcus phage S-SBP1 (MOI=2). Samples were taken 15min, 1h, 3h, 5h and 7h post-infection (n=1). [GSE150732]'\n",
    "doi = 'https://doi.org/10.1002/mbo3.1150'\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Huang_2021/notebook/Huang_infection_fractional_expression.tsv', norm = 'fractional',\n",
    "               name = 'Haung_2021_infection', year = 2021, journal = 'Microbiology Open', firstauthor = 'Huang', pubmedID = 33377630, description = description, host_id=21, phage_id=22, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Huang_2021/notebook/Huang_infection_full_TPM.tsv', norm = 'TPM',\n",
    "               name = 'Haung_2021_infection', year = 2021, journal = 'Microbiology Open', firstauthor = 'Huang', pubmedID = 33377630, description = description, host_id=21, phage_id=22, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Huang_2021/notebook/Huang_infection_TPM_means.tsv', norm = 'TPM_means',\n",
    "               name = 'Haung_2021_infection', year = 2021, journal = 'Microbiology Open', firstauthor = 'Huang', pubmedID = 33377630, description = description, host_id=21, phage_id=22, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Huang_2021/notebook/Huang_infection_TPM_std.tsv', norm = 'TPM_std',\n",
    "               name = 'Haung_2021_infection', year = 2021, journal = 'Microbiology Open', firstauthor = 'Huang', pubmedID = 33377630, description = description, host_id=21, phage_id=22, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Huang_2021/notebook/Huang_infection_full_raw_counts.tsv', norm = 'raw_counts',\n",
    "               name = 'Haung_2021_infection', year = 2021, journal = 'Microbiology Open', firstauthor = 'Huang', pubmedID = 33377630, description = description, host_id=21, phage_id=22, doi = doi, upload_date = '25-08-12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "id": "23117dcd",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [38] Add Furi dataset\n",
    "\n",
    "description = 'Infection of Streptococcus pneumoniae strain D39 with Streptococcus phage SpSL1 (MOI=0.2). Samples were taken 10, 50, 90 min post-infection in biological triplicates (n=3). [GSE132611]'\n",
    "doi = 'https://doi.org/10.1128/jb.00370-19'\n",
    "\n",
    "with app.app_context():\n",
    "    addFullDataset(phageName='SpSL1', phageDes = 'Streptococcus phage SpSL1', phageID = 'KM882824.1', phageType = 'temperate',\n",
    "               hostName = 'S. pneumoniae strain D39', hostDes = 'Streptococcus pneumoniae strain D39', hostID = 'NC_008533.2', hostGroup = 'Streptococcus pneumoniae',\n",
    "               rnaSeqPath = 'Datasets/Furi_2019/notebook/Furi_fractional_expression.tsv', norm = 'fractional',\n",
    "               name = 'Furi_2019', year = 2019, journal = 'mSphere', firstauthor = 'Furi', pubmedID = 31285240, description = description, doi = doi, upload_date = '25-08-12',\n",
    "               phageGenomePath = 'Datasets/Furi_2019/notebook/Furi_phage_gff3.tsv', phageGenomeName = 'Streptococcus phage SpSL1 genome',\n",
    "               hostGenomePath = 'Datasets/Furi_2019/notebook/Furi_host_gff3.tsv', hostGenomeName = 'S. pneumoniae strain D39 genome')\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Furi_2019/notebook/Furi_full_TPM.tsv', norm = 'TPM',\n",
    "               name = 'Furi_2019', year = 2019, journal = 'mSphere', firstauthor = 'Furi', pubmedID = 31285240, description = description, host_id=22, phage_id=23, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Furi_2019/notebook/Furi_TPM_means.tsv', norm = 'TPM_means',\n",
    "               name = 'Furi_2019', year = 2019, journal = 'mSphere', firstauthor = 'Furi', pubmedID = 31285240, description = description, host_id=22, phage_id=23, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Furi_2019/notebook/Furi_TPM_std.tsv', norm = 'TPM_std',\n",
    "               name = 'Furi_2019', year = 2019, journal = 'mSphere', firstauthor = 'Furi', pubmedID = 31285240, description = description, host_id=22, phage_id=23, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Furi_2019/notebook/Furi_full_raw_counts.tsv', norm = 'raw_counts',\n",
    "               name = 'Furi_2019', year = 2019, journal = 'mSphere', firstauthor = 'Furi', pubmedID = 31285240, description = description, host_id=22, phage_id=23, doi = doi, upload_date = '25-08-12')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "d7c1b808",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [39] Add Doron WH7803 dataset\n",
    "\n",
    "description = 'Infection of Synechococcus sp. WH7803 with T4-like Synechococcus phage syn9. Samlpes were taken 5, 30, 60, 120 min post-infection in biological duplicates (n=2). [GSE74921]'\n",
    "doi = 'https://doi.org/10.1038/ismej.2015.210'\n",
    "\n",
    "with app.app_context():\n",
    "    phageAdd = Phage(name = 'Syn9', description = 'Synechococcus phage Syn9', ncbi_id = 'NC_008296.2', phage_type = 'virulent')\n",
    "    db.session.add(phageAdd)\n",
    "    db.session.commit()\n",
    "\n",
    "with app.app_context():\n",
    "    phagegff = pd.read_csv('Datasets/Doron_2016/notebook/Doron_WH7803_phage_gff3.tsv', sep='\\t', comment = '#')\n",
    "    phagegenome = pickle.dumps(phagegff)\n",
    "    phageGFF = PhageGenome(name = 'Synechoccus phage Syn9 genome', phage_id = 24, gff_data = phagegenome)\n",
    "    db.session.add(phageGFF)\n",
    "    db.session.commit()\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Doron_2016/notebook/Doron_WH7803_fractional_expression.tsv', norm = 'fractional',\n",
    "               name = 'Doron_2016_WH7803', year = 2016, journal = 'ISME', firstauthor = 'Doron', pubmedID = 26623542, description = description, host_id=21, phage_id=24, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Doron_2016/notebook/Doron_WH7803_full_TPM.tsv', norm = 'TPM',\n",
    "               name = 'Doron_2016_WH7803', year = 2016, journal = 'ISME', firstauthor = 'Doron', pubmedID = 26623542, description = description, host_id=21, phage_id=24, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Doron_2016/notebook/Doron_WH7803_TPM_means.tsv', norm = 'TPM_means',\n",
    "               name = 'Doron_2016_WH7803', year = 2016, journal = 'ISME', firstauthor = 'Doron', pubmedID = 26623542, description = description, host_id=21, phage_id=24, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Doron_2016/notebook/Doron_WH7803_TPM_std.tsv', norm = 'TPM_std',\n",
    "               name = 'Doron_2016_WH7803', year = 2016, journal = 'ISME', firstauthor = 'Doron', pubmedID = 26623542, description = description, host_id=21, phage_id=24, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Doron_2016/notebook/Doron_WH7803_full_raw_counts.tsv', norm = 'raw_counts',\n",
    "               name = 'Doron_2016_WH7803', year = 2016, journal = 'ISME', firstauthor = 'Doron', pubmedID = 26623542, description = description, host_id=21, phage_id=24, doi = doi, upload_date = '25-08-12')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "id": "0701e061",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [40] Add Doron WH8102 dataset\n",
    "\n",
    "description = 'Infection of Synechococcus sp. WH8102 with T4-like Synechococcus phage syn9. Samlpes were taken 5, 30, 60, 120 min post-infection in biological duplicates (n=2). [GSE74921]'\n",
    "doi = 'https://doi.org/10.1038/ismej.2015.210'\n",
    "\n",
    "with app.app_context():\n",
    "    hostAdd = Host(name = 'Synechococcus WH8102', description = 'Synechococcus sp. WH8102', ncbi_id = 'NC_005070', group = 'Synechococcus species')\n",
    "    db.session.add(hostAdd)\n",
    "    db.session.commit()\n",
    "\n",
    "with app.app_context():\n",
    "    hostgff = pd.read_csv('Datasets/Doron_2016/notebook/Doron_WH8102_host_gff3.tsv', sep='\\t', comment = '#')\n",
    "    hostgenome = pickle.dumps(hostgff)\n",
    "    hostGFF = HostGenome(name = 'Synechococcus sp. WH8102 genome', host_id = 23, gff_data = hostgenome)\n",
    "    db.session.add(hostGFF)\n",
    "    db.session.commit()\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Doron_2016/notebook/Doron_WH8102_fractional_expression.tsv', norm = 'fractional',\n",
    "               name = 'Doron_2016_WH8102', year = 2016, journal = 'ISME', firstauthor = 'Doron', pubmedID = 26623542, description = description, host_id=23, phage_id=24, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Doron_2016/notebook/Doron_WH8102_full_TPM.tsv', norm = 'TPM',\n",
    "               name = 'Doron_2016_WH8102', year = 2016, journal = 'ISME', firstauthor = 'Doron', pubmedID = 26623542, description = description, host_id=23, phage_id=24, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Doron_2016/notebook/Doron_WH8102_TPM_means.tsv', norm = 'TPM_means',\n",
    "               name = 'Doron_2016_WH8102', year = 2016, journal = 'ISME', firstauthor = 'Doron', pubmedID = 26623542, description = description, host_id=23, phage_id=24, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Doron_2016/notebook/Doron_WH8102_TPM_std.tsv', norm = 'TPM_std',\n",
    "               name = 'Doron_2016_WH8102', year = 2016, journal = 'ISME', firstauthor = 'Doron', pubmedID = 26623542, description = description, host_id=23, phage_id=24, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Doron_2016/notebook/Doron_WH8102_full_raw_counts.tsv', norm = 'raw_counts',\n",
    "               name = 'Doron_2016_WH8102', year = 2016, journal = 'ISME', firstauthor = 'Doron', pubmedID = 26623542, description = description, host_id=23, phage_id=24, doi = doi, upload_date = '25-08-12')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "id": "d5b21042",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [41] Add Doron WH8109 dataset\n",
    "\n",
    "description = 'Infection of Synechococcus sp. WH8109 with T4-like Synechococcus phage syn9. Samlpes were taken 5, 30, 60, 120 min post-infection in biological duplicates (n=2). [GSE74921]'\n",
    "doi = 'https://doi.org/10.1038/ismej.2015.210'\n",
    "\n",
    "with app.app_context():\n",
    "    hostAdd = Host(name = 'Synechococcus WH8109', description = 'Synechococcus sp. WH8109', ncbi_id = 'CP006882', group = 'Synechococcus species')\n",
    "    db.session.add(hostAdd)\n",
    "    db.session.commit()\n",
    "\n",
    "with app.app_context():\n",
    "    hostgff = pd.read_csv('Datasets/Doron_2016/notebook/Doron_WH8109_host_gff3.tsv', sep='\\t', comment = '#')\n",
    "    hostgenome = pickle.dumps(hostgff)\n",
    "    hostGFF = HostGenome(name = 'Synechococcus sp. WH8109 genome', host_id = 24, gff_data = hostgenome)\n",
    "    db.session.add(hostGFF)\n",
    "    db.session.commit()\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Doron_2016/notebook/Doron_WH8109_fractional_expression.tsv', norm = 'fractional',\n",
    "               name = 'Doron_2016_WH8109', year = 2016, journal = 'ISME', firstauthor = 'Doron', pubmedID = 26623542, description = description, host_id=24, phage_id=24, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Doron_2016/notebook/Doron_WH8109_full_TPM.tsv', norm = 'TPM',\n",
    "               name = 'Doron_2016_WH8109', year = 2016, journal = 'ISME', firstauthor = 'Doron', pubmedID = 26623542, description = description, host_id=24, phage_id=24, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Doron_2016/notebook/Doron_WH8109_TPM_means.tsv', norm = 'TPM_means',\n",
    "               name = 'Doron_2016_WH8109', year = 2016, journal = 'ISME', firstauthor = 'Doron', pubmedID = 26623542, description = description, host_id=24, phage_id=24, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Doron_2016/notebook/Doron_WH8109_TPM_std.tsv', norm = 'TPM_std',\n",
    "               name = 'Doron_2016_WH8109', year = 2016, journal = 'ISME', firstauthor = 'Doron', pubmedID = 26623542, description = description, host_id=24, phage_id=24, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Doron_2016/notebook/Doron_WH8109_full_raw_counts.tsv', norm = 'raw_counts',\n",
    "               name = 'Doron_2016_WH8109', year = 2016, journal = 'ISME', firstauthor = 'Doron', pubmedID = 26623542, description = description, host_id=24, phage_id=24, doi = doi, upload_date = '25-08-12')\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "id": "3d70fa40",
   "metadata": {},
   "outputs": [],
   "source": [
    "# [42] Add Halleran dataset\n",
    "\n",
    "description = 'Cluster A4 mycobacteriophage infecting Mycobacterium smegmatis (model organism for mycobacterium tubercolosis) at MOI=10. Samples were taken before (0) and 5, 15, 30, 60, 120 min post-infection in biological duplicates (n=2). [GSE60592]'\n",
    "doi = 'https://doi.org/10.1371/journal.pone.0141100'\n",
    "\n",
    "with app.app_context():\n",
    "    addFullDataset(phageName='Kampy', phageDes = 'A4 Mycobacterium phage Kampy', phageID = 'KJ510414.1', phageType = 'virulent',\n",
    "               hostName = 'M. smegmatis strain mc2 155', hostDes = 'Mycobacterium smegmatis strain mc2 155', hostID = 'NC_018289.1', hostGroup = 'Mycobacterium smegmatis',\n",
    "               rnaSeqPath = 'Datasets/Halleran_2015/notebook/Halleran_fractional_expression.tsv', norm = 'fractional',\n",
    "               name = 'Halleran_2015', year = 2015, journal = 'PLOS One', firstauthor = 'Halleran', pubmedID = 26513661, description = description, doi = doi, upload_date = '25-08-12',\n",
    "               phageGenomePath = 'Datasets/Halleran_2015/notebook/Halleran_phage_gff3.tsv', phageGenomeName = 'A4 Mycobacterium phage Kampy genome',\n",
    "               hostGenomePath = 'Datasets/Halleran_2015/notebook/Halleran_host_gff3.tsv', hostGenomeName = 'M. smegmatis strain mc2 155 genome')\n",
    "\n",
    "with app.app_context():\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Halleran_2015/notebook/Halleran_full_TPM.tsv', norm = 'TPM',\n",
    "               name = 'Halleran_2015', year = 2015, journal = 'PLOS One', firstauthor = 'Halleran', pubmedID = 26513661, description = description, host_id=25, phage_id=25, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Halleran_2015/notebook/Halleran_TPM_means.tsv', norm = 'TPM_means',\n",
    "               name = 'Halleran_2015', year = 2015, journal = 'PLOS One', firstauthor = 'Halleran', pubmedID = 26513661, description = description, host_id=25, phage_id=25, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Halleran_2015/notebook/Halleran_TPM_std.tsv', norm = 'TPM_std',\n",
    "               name = 'Halleran_2015', year = 2015, journal = 'PLOS One', firstauthor = 'Halleran', pubmedID = 26513661, description = description, host_id=25, phage_id=25, doi = doi, upload_date = '25-08-12')\n",
    "    addSingleDataset(rnaSeqPath = 'Datasets/Halleran_2015/notebook/Halleran_full_raw_counts.tsv', norm = 'raw_counts',\n",
    "               name = 'Halleran_2015', year = 2015, journal = 'PLOS One', firstauthor = 'Halleran', pubmedID = 26513661, description = description, host_id=25, phage_id=25, doi = doi, upload_date = '25-08-12')"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "PhageExpAtlas",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.12.12"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
