{% extends "index.html" %}

{% block title %}Help{% endblock %}

{% block content %}
<div id="content-help">

    <div class="page-header">
        <h1>Help & Information</h1>
    </div>

    
    <div id="help-main-content" class="main-content">
        
        <sl-details summary="Getting started" open>
            Welcome to the PhageExpressionAtlas, your platform to explore various phage infections on the transcriptional level based on dual RNA-seq data. Here, we provide some general guidelines to work with the atlas and additional info. <br><br>
            <ul>
                <li>In the current implementation, the PhageExpressionAtlas provides uniformly processed dual RNA-seq data of diverse phage infections for exploration and visualization.</li>
                <li>To find out more about the datasets that are currently hosted on the PhageExpressionAtlas visit <a href="/data-overview">Data Overview</a>.</li>
                <li>In order to explore the datasets currently hosted in the PhageExpressionAtlas you may visit the <a href="/dataset-exploration">Dataset Exploration</a> page. There, select your dataset and/or phage and host of interest to investigate host and phage gene expression on the transcriptional level. For instance, phage gene classification (early, middle, late) can be obtained and visualized in the genome viewer at their respective genomic location.</li>
            </ul>
            
        </sl-details>


        <sl-details summary="Guide" open>
            
            <!-- Home -->
            <a href="/">
                <div class="heading-wrapper underline">
                    <svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
                        <path fill-rule="evenodd" clip-rule="evenodd" d="M12.7142 4.30057C12.322 3.89981 11.677 3.89981 11.2848 4.30057L4.32831 11.409C3.70882 12.0421 4.15731 13.1085 5.04302 13.1085H5.87066V18.9677C5.87066 19.5213 6.33758 19.9701 6.91355 19.9701H9.39246C9.96844 19.9701 10.4354 19.5213 10.4354 18.9677V16.9629C10.4354 16.4093 10.9023 15.9605 11.4783 15.9605H12.5211C13.0971 15.9605 13.564 16.4093 13.564 16.9629V18.9677C13.564 19.5213 14.031 19.9701 14.6069 19.9701H17.0858C17.6618 19.9701 18.1287 19.5213 18.1287 18.9677V13.1085H18.956C19.8417 13.1085 20.2902 12.0421 19.6707 11.409L12.7142 4.30057Z" fill="#1D1B20"></path>
                    </svg>
                    <strong>Home</strong>
                </div>
            </a>
            <br>
            <p>Get an overview of the PhageExpressionAtlas and the possibilities it offers. Additionally, find background information about bacteriophages and dual RNA-sequencing. </p>
            
            <br>
            <sl-divider></sl-divider>
            <br>

            <!-- Data Overview -->
            <a href="/data-overview">
                <div class="heading-wrapper underline" id="guide-overview">
                    <svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
                        <path fill-rule="evenodd" clip-rule="evenodd" d="M18.296 6.7544C18.296 6.78757 18.2946 6.8206 18.292 6.85349H18.3043V6.86464C18.2623 7.04484 18.15 7.24304 17.9364 7.45521C17.6519 7.73767 17.2149 8.01505 16.6347 8.2587C15.4765 8.74514 13.8417 9.05792 12.0099 9.05792C10.1782 9.05792 8.54337 8.74514 7.38513 8.2587C6.80498 8.01505 6.36794 7.73767 6.08348 7.45521C5.81325 7.18689 5.70522 6.94091 5.69629 6.72542C5.73184 5.21756 8.53868 4 11.996 4C15.4754 4 18.296 5.23319 18.296 6.7544ZM18.3043 8.45822C17.9459 8.73476 17.5096 8.97587 17.0219 9.18069C15.7139 9.73004 13.9416 10.0579 12.0099 10.0579C10.0783 10.0579 8.30595 9.73004 6.99791 9.18069C6.50516 8.97374 6.06486 8.72974 5.70433 8.44957V11.1453C6.0389 11.4064 6.51379 11.6592 7.10957 11.883C8.42273 12.3762 10.2191 12.6893 12.0844 12.6893C13.9497 12.6893 15.7016 12.3762 16.9698 11.8848C17.5364 11.6653 17.9861 11.418 18.3043 11.1631V8.45822ZM18.3043 12.3622C18.0092 12.5297 17.6817 12.6815 17.3311 12.8173C15.926 13.3617 14.0476 13.6893 12.0844 13.6893C10.1214 13.6893 8.20273 13.3617 6.75798 12.8191C6.3773 12.6762 6.02218 12.5155 5.70433 12.3376V14.7762C6.0389 15.0373 6.51379 15.2901 7.10957 15.5139C8.42273 16.0071 10.2191 16.3202 12.0844 16.3202C13.9497 16.3202 15.7016 16.0071 16.9698 15.5157C17.5364 15.2962 17.9861 15.0489 18.3043 14.794V12.3622ZM5.70433 17.3782V15.9685C6.02218 16.1464 6.3773 16.3071 6.75798 16.45C8.20273 16.9927 10.1214 17.3202 12.0844 17.3202C14.0476 17.3202 15.926 16.9926 17.3311 16.4482C17.6817 16.3124 18.0092 16.1606 18.3043 15.9931V17.3782V17.4798H18.3C18.1759 18.9299 15.4047 20.0891 12.0043 20.0891C8.60398 20.0891 5.83282 18.9299 5.70867 17.4798H5.70433V17.3782Z" fill="#1D1B20"></path>
                    </svg>
                    <strong>Data Overview</strong>
                </div>
            </a>
            <br>
            <p>Take a look at how the data was processed, which phages and hosts are represented in the data and which datasets are available. </p>

            <br>
            <p>
                <strong>
                    Data visualizations in Data Overview:
                </strong>
            </p>

            <br>

            <div class="column-wrapper">
                <sl-carousel id="overview-dataviz-carousel" loop navigation pagination>
                    <sl-carousel-item>
                      <img
                        alt="Sankey Diagram Example of the overview of all currenttly available Phage-Host Interactions in the PhageExpressionAtlas"
                        src="/static/assets/help/overview-dataviz-carousel/sankey.png"
                      />
                      <p>Example: Overview of available Phage-Host Interactions with arrow pointing to the download button</p>
                    </sl-carousel-item>
                    <sl-carousel-item>
                      <img
                        alt="Pie Chart Example of Phage Lifestyles represented in the PhageExpressionAtlas"
                        src="/static/assets/help/overview-dataviz-carousel/phagetype.png"
                      />
                      <p>Example: Proportional Distribution of Phage Lifestyles with tooltip</p>
                    </sl-carousel-item>
                    <sl-carousel-item>
                      <img
                        alt="Pie Chart Example of Phage Distribution"
                        src="/static/assets/help/overview-dataviz-carousel/phagedistribution.png"
                      />

                      <p>Example: Proportional Distribution of Phages across all Datasets</p>

                    </sl-carousel-item>
                    <sl-carousel-item>
                      <img
                        alt="Sunburst Chart Example of Host (strains) represented in the Atlas"
                        src="/static/assets/help/overview-dataviz-carousel/hostdistribution.png"
                      />

                      <p>Example: Proportional Distribution of Hosts across all Datasets</p>
                      
                    </sl-carousel-item>
                
                </sl-carousel>

                <div class="text-wrapper">

                    <p>Information about the distribution of the PhageExpressionAtlas, such as which phages are currently represented in the atlas, can be explored visually and interactively with the different charts (examples of the charts are shown on the left side).</p>

                    <br>

                    <p>All Charts can be downloaded as png by click the  "<sl-icon name="camera-fill" class="light-gray"></sl-icon>"  button above each chart <span class="slide-reference" data-slide="1" data-link="overview-dataviz-carousel">(see slide 1)</span>.</p>

                    <br>
                    <p>To explore the charts in more detail, simply hover over any part of it. A tooltip will appear showing helpful information such as the percentage, or value of each category <span class="slide-reference" data-slide="2" data-link="overview-dataviz-carousel">(see slide 2)</span>.
                    </p>
                    <br>
                    <p>To hide or display categories in the pie charts, it is possible to click the legend entries above the charts which look like this: 
                        <div class="legend-entry">
                            <div class="rectangle col5"></div> Category 1..
                        </div>
                        &nbsp;&nbsp;
                        <div class="legend-entry">
                            <div class="rectangle col3"></div> Category 2..
                        </div>
                    </p>
                    <br>
                    <p>
                        Some legends also include pagination. The pages can be navigated by clicking on "<sl-icon name="caret-left-fill" class="light-gray"></sl-icon>" and "<sl-icon name="caret-right-fill" class="light-gray"></sl-icon>"which are located next to the legend <span class="slide-reference" data-slide="3" data-link="overview-dataviz-carousel">(see slide 3)</span>.
                    </p>
                    <br>
                    <p>
                        In the sunburst plot <span class="slide-reference" data-slide="4" data-link="overview-dataviz-carousel">(see slide 4)</span> categories can be selected by clicking. To show all categories again, the category has to be clicked again.
                    </p>
                    
                    
                    
                </div>
            </div>
            
            <br>
            <p>
                <strong>
                    Where to find all Datasets currently in the Atlas:
                </strong>
            </p>
            <p>
                All datasets currently available for exploration in the Atlas can be found on the Data Overview page, under the <a href="/data-overview#datasets">Datasets section</a>. 
                The video on the right provides a visual guide for the table, where each dataset is represented as a row.
            </p>
            <br>

            <div class="column-wrapper">
                <div class="text-wrapper">
                
                    <p class="underline">How to navigate the table?</p>
                    <p>
                        The table contains a variety of details for each dataset — such as the study name, doi, and PubMed id — making it quite long. To view all the information, it is needed to scroll through the table (as shown in the video on the right).
                    </p>
                    <br>  
                    <p>
                        Additionally, the table is paginated. You can navigate between pages using the buttons located below the table.
                        You can also control how many rows are shown at once by adjusting the page size.
                    </p>
                    
                    <p>
                        The table also offers the option to view a specific dataset in either dataset exploration or the genome viewer, by clicking either "<svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
                            <path fill-rule="evenodd" clip-rule="evenodd" d="M8 3.98438H14.1841L18 7.80029V17.9844C18 19.0889 17.1046 19.9844 16 19.9844H8C6.89543 19.9844 6 19.0889 6 17.9844V5.98438C6 4.87981 6.89543 3.98438 8 3.98438ZM14.1797 4.68604V6.80464C14.1797 7.35693 14.6274 7.80464 15.1797 7.80464H17.2744L14.2978 4.82806C14.2534 4.78362 14.214 4.73604 14.1797 4.68604ZM11.313 14.6626C12.105 14.6626 12.7471 14.0205 12.7471 13.2285C12.7471 12.4365 12.105 11.7944 11.313 11.7944C10.521 11.7944 9.87891 12.4365 9.87891 13.2285C9.87891 14.0205 10.521 14.6626 11.313 14.6626ZM11.313 15.6626C11.8309 15.6626 12.311 15.5009 12.7056 15.2251L14.3475 16.8671C14.5246 17.0442 14.8118 17.0442 14.9889 16.8671C15.166 16.69 15.166 16.4028 14.9889 16.2257L13.3399 14.5767C13.5971 14.1908 13.7471 13.7271 13.7471 13.2285C13.7471 11.8842 12.6573 10.7944 11.313 10.7944C9.96868 10.7944 8.87891 11.8842 8.87891 13.2285C8.87891 14.5728 9.96868 15.6626 11.313 15.6626Z" fill="#1D1B20"></path>
                        </svg>" or "<svg width="25" height="24" viewBox="0 0 25 24" fill="none" xmlns="http://www.w3.org/2000/svg">
                            <path d="M16.0033 19.4757L16.0026 19.4633C16.0019 19.4522 16.0008 19.4345 15.9992 19.4106C15.9958 19.3626 15.9903 19.2898 15.9819 19.1956C15.9649 19.0071 15.9361 18.7344 15.8894 18.4058C15.8666 18.2461 15.8398 18.0744 15.8083 17.8938H8.71294V16.3938H15.4668C15.4311 16.27 15.3932 16.1458 15.3529 16.0219C15.3197 15.9201 15.2853 15.8193 15.2495 15.72H8.71294V14.22H14.5176C14.4205 14.0758 14.3183 13.9414 14.2108 13.8185C13.7485 13.2899 13.2181 13 12.547 13C11.179 13 10.133 12.3618 9.37768 11.4982C8.64317 10.6584 8.16138 9.58702 7.83921 8.59689C7.51344 7.59567 7.32849 6.60665 7.22452 5.87599C7.17223 5.50849 7.1397 5.20115 7.12013 4.98362C7.11033 4.87475 7.10375 4.78806 7.09955 4.72723C7.09745 4.69681 7.09594 4.67284 7.09492 4.65576L7.09375 4.63536L7.09341 4.62914L7.09331 4.62705C7.09329 4.62673 7.09323 4.62564 8.09195 4.57495L7.09323 4.62564C7.06524 4.07406 7.48968 3.60423 8.04126 3.57624C8.59257 3.54826 9.06222 3.97228 9.09062 4.52347L9.09072 4.5253L9.09138 4.53674C9.09204 4.54776 9.09314 4.56548 9.0948 4.58944C9.09811 4.63738 9.10361 4.71022 9.11208 4.80437C9.12905 4.9929 9.15782 5.26565 9.20458 5.59424C9.2273 5.75393 9.2541 5.92561 9.2856 6.1062H17.0054V7.6062H9.62716C9.66282 7.72997 9.70075 7.85415 9.74107 7.97808C9.77421 8.07993 9.80867 8.18067 9.84447 8.28003H16.3942V9.78003H10.5763C10.6734 9.92425 10.7756 10.0586 10.8831 10.1815C11.3454 10.7101 11.8758 11 12.547 11C13.9149 11 14.9609 11.6382 15.7163 12.5018C16.4508 13.3416 16.9326 14.413 17.2547 15.4031C17.5805 16.4043 17.7655 17.3934 17.8694 18.124C17.9217 18.4915 17.9542 18.7988 17.9738 19.0164C17.9836 19.1253 17.9902 19.2119 17.9944 19.2728C17.9965 19.3032 17.998 19.3272 17.999 19.3442L18.0002 19.3646L18.0005 19.3709L18.0006 19.3729C18.0007 19.3733 18.0007 19.3744 17.002 19.425L18.0007 19.3744C18.0287 19.9259 17.6043 20.3958 17.0527 20.4238C16.5014 20.4517 16.0317 20.0269 16.0033 19.4757Z" fill="#1D1B20"></path>
                            <path d="M9.09053 19.4755L9.09059 19.4745L9.09125 19.463C9.09191 19.452 9.09302 19.4343 9.09467 19.4103C9.09798 19.3624 9.10348 19.2896 9.11195 19.1954C9.12892 19.0069 9.1577 18.7341 9.20445 18.4055C9.29859 17.744 9.46274 16.8767 9.74095 16.0217C10.0228 15.1556 10.3998 14.3707 10.883 13.8183C11.3453 13.2897 11.8758 13 12.547 13C13.9149 13 14.9608 12.3616 15.7161 11.498C16.4507 10.6582 16.9324 9.5868 17.2546 8.59667C17.5804 7.59545 17.7653 6.60643 17.8693 5.87577C17.9216 5.50827 17.9541 5.20093 17.9737 4.9834C17.9835 4.87453 17.9901 4.78784 17.9943 4.72701C17.9964 4.69659 17.9979 4.67262 17.9989 4.65554L18.0001 4.63514L18.0004 4.62892L18.0005 4.62683C18.0005 4.62651 18.0006 4.62542 17.0019 4.57473L18.0006 4.62542C18.0286 4.07384 17.6041 3.60401 17.0526 3.57602C16.5013 3.54804 16.0316 3.97206 16.0032 4.52325C16.0032 4.52327 16.0032 4.52323 16.0032 4.52325L16.0031 4.52508L16.0024 4.53652C16.0018 4.54754 16.0007 4.56526 15.999 4.58922C15.9957 4.63716 15.9902 4.71 15.9817 4.80415C15.9648 4.99268 15.936 5.26543 15.8892 5.59402C15.7951 6.25557 15.631 7.12282 15.3528 7.97786C15.0709 8.844 14.6939 9.62889 14.2107 10.1813C13.7484 10.7099 13.2181 11 12.547 11C11.179 11 10.1329 11.638 9.37755 12.5016C8.64305 13.3414 8.16125 14.4128 7.83909 15.4029C7.51331 16.4041 7.32837 17.3931 7.2244 18.1238C7.1721 18.4913 7.13958 18.7986 7.12 19.0162C7.11021 19.125 7.10363 19.2117 7.09942 19.2725C7.09732 19.303 7.09582 19.3269 7.0948 19.344L7.09362 19.3644L7.09329 19.3706L7.09318 19.3727C7.09316 19.3731 7.09311 19.3741 8.09182 19.4248L7.09311 19.3741C7.06511 19.9257 7.48956 20.3955 8.04113 20.4235C8.59244 20.4515 9.06213 20.0267 9.09053 19.4755Z" fill="#1D1B20"></path>
                        </svg>" in the corresponding dataset row. An example on how to navigate the table can be seen in <span class="slide-reference" data-slide="1" data-link="overview-table-carousel">slide 1</span>.
                    </p>

                    <br>
                    <p class="underline">Is it possible to filter the table?</p>
                    <p>Yes, above the table is the filter button located. By clicking on "<sl-icon name="funnel-fill" class="col3"></sl-icon>", the filter section will be opened. In the first selection element the column of interest can be selected and in the second select the desired value can be choosen. After making your selections, the table will update accordingly.
                    To clear the filter you can click on the "Clear Filter" button next to the selection elements.</p>
                    
                    <p>Furthermore, each column can be sorted either in descending or ascending order by clicking on "<sl-icon name="caret-up-fill" class="light-gray"></sl-icon>" next to the respective column name. An example on how to filter the table can be seen in <span class="slide-reference" data-slide="2" data-link="overview-table-carousel">slide 2</span>.</p>
                    
                    <br>

                    <p class="underline">Can I download the table?</p>
                    <p>Also yes, the download button is located above the table, adjacent to the filter button. Click "<sl-icon name="download" class="col3"></sl-icon>" to download the table in CSV format.</p>
                    
                </div>

                <sl-carousel id="overview-table-carousel" loop navigation pagination>
                    <sl-carousel-item>
                        <video controls autoplay muted loop aria-label="Video of dataset table that shows how to navigate the table">
                            <source src="static/assets/help/overview-table-carousel/example_table_navigation.mp4" type="video/mp4">
                            Your browser does not support the video tag.
                        </video>
                        <p>Example: How to navigate the table</p>
                    </sl-carousel-item>
                    <sl-carousel-item>
                        <video controls autoplay muted loop aria-label="Video of dataset table that shows how to filter the table">
                            <source src="static/assets/help/overview-table-carousel/example_table_filtering.mp4" type="video/mp4">
                            Your browser does not support the video tag.
                        </video>
                        <p>Example: How to filter the table</p>
                    </sl-carousel-item>
                    
                    
                
                </sl-carousel>

            </div>

            <br>
            <sl-divider></sl-divider>
            <br>

            <!-- Dataset Exploration -->
            <a href="/dataset-exploration">
                <div class="heading-wrapper underline" id="guide-exploration">
                    <svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
                        <path fill-rule="evenodd" clip-rule="evenodd" d="M8 3.98438H14.1841L18 7.80029V17.9844C18 19.0889 17.1046 19.9844 16 19.9844H8C6.89543 19.9844 6 19.0889 6 17.9844V5.98438C6 4.87981 6.89543 3.98438 8 3.98438ZM14.1797 4.68604V6.80464C14.1797 7.35693 14.6274 7.80464 15.1797 7.80464H17.2744L14.2978 4.82806C14.2534 4.78362 14.214 4.73604 14.1797 4.68604ZM11.313 14.6626C12.105 14.6626 12.7471 14.0205 12.7471 13.2285C12.7471 12.4365 12.105 11.7944 11.313 11.7944C10.521 11.7944 9.87891 12.4365 9.87891 13.2285C9.87891 14.0205 10.521 14.6626 11.313 14.6626ZM11.313 15.6626C11.8309 15.6626 12.311 15.5009 12.7056 15.2251L14.3475 16.8671C14.5246 17.0442 14.8118 17.0442 14.9889 16.8671C15.166 16.69 15.166 16.4028 14.9889 16.2257L13.3399 14.5767C13.5971 14.1908 13.7471 13.7271 13.7471 13.2285C13.7471 11.8842 12.6573 10.7944 11.313 10.7944C9.96868 10.7944 8.87891 11.8842 8.87891 13.2285C8.87891 14.5728 9.96868 15.6626 11.313 15.6626Z" fill="#1D1B20"></path>
                    </svg>
                    <strong>Dataset Exploration</strong>
                </div>
            </a>
            <br>
            <p>Explore a dataset of your interest. Inspect overall patterns of host and phage transcription during infection, phage gene classes or study the expression of particular host or phage genes of interest.</p>

            <br>
            <p class="underline">Where to start?</p>
            <p>
                After opening the Dataset Exploration page, a specific phage, host and study are already pre-selected in the selection elements on the left. To make your own selections, first select a phage, then a host, and finally the study. Once everything is selected, the visualizations and information will be updated.
            </p>

            <p>
                <strong>Important:</strong> When making your selections, make sure you start from the top down, as the selections are mostly interdependent. So no worries, if nothing is selected, the host and study selection items will be empty, so just select your phage of interest and you are good to go.
            </p>

            <br>
            <p class="underline">
                You can not find your study of interest in the select element?
            </p>
            <p> 
                As mentioned above, the selections are interdependent, so it is possible that you have selected the wrong phage or host and the study is not listed. Feel free to click the "Deselect all" button and start over.
            </p>

            <br>
            <p class="underline">
                You want to look at a specific dataset and don't know which phage or host to choose?
            </p>
            <p>
                No worries, just look at the dataset table on the Data Overview page, there is a specific button for each dataset ("<svg width="24" height="24" viewBox="0 0 24 24" fill="none" xmlns="http://www.w3.org/2000/svg">
                    <path fill-rule="evenodd" clip-rule="evenodd" d="M8 3.98438H14.1841L18 7.80029V17.9844C18 19.0889 17.1046 19.9844 16 19.9844H8C6.89543 19.9844 6 19.0889 6 17.9844V5.98438C6 4.87981 6.89543 3.98438 8 3.98438ZM14.1797 4.68604V6.80464C14.1797 7.35693 14.6274 7.80464 15.1797 7.80464H17.2744L14.2978 4.82806C14.2534 4.78362 14.214 4.73604 14.1797 4.68604ZM11.313 14.6626C12.105 14.6626 12.7471 14.0205 12.7471 13.2285C12.7471 12.4365 12.105 11.7944 11.313 11.7944C10.521 11.7944 9.87891 12.4365 9.87891 13.2285C9.87891 14.0205 10.521 14.6626 11.313 14.6626ZM11.313 15.6626C11.8309 15.6626 12.311 15.5009 12.7056 15.2251L14.3475 16.8671C14.5246 17.0442 14.8118 17.0442 14.9889 16.8671C15.166 16.69 15.166 16.4028 14.9889 16.2257L13.3399 14.5767C13.5971 14.1908 13.7471 13.7271 13.7471 13.2285C13.7471 11.8842 12.6573 10.7944 11.313 10.7944C9.96868 10.7944 8.87891 11.8842 8.87891 13.2285C8.87891 14.5728 9.96868 15.6626 11.313 15.6626Z" fill="#1D1B20"></path>
                </svg>") that when clicked will redirect you to the Dataset Exploration page with your study of interest already selected (for more information on the table, please see the Data Overview Guide).
            </p>

            <p class="underline">
                Can i search in the select elements?
            </p>
            <p>
                As of now, yes and no. When the select is opened and all options are visible, it is possible to enter the starting letters of the option you are looking for. When entering the option window jumps to the first option that matches the letters you have entered. When clicking 'Enter' on your keyboard this first option is being selected. So while you can search, the function is not perfect at the moment. 
            </p>

            <br>
            <p>
                <strong>
                    Data visualizations in Dataset Exploration:
                </strong>
            </p>
            <p>The Atlas offers several visualizations to help you explore phage-host interactions and phage gene classifications within the selected dataset. The primary visual tools include heatmaps and gene expression profiles over time.</p>
            <p>These visualizations are organized into two main sections:</p>
            <ul>
                <li>Top Section – Phage-Host Interaction Overview 
                    <br>
                    This area provides a high-level summary of the interactions between phages and their hosts, giving you a quick overview of the dataset.
                </li>
                
                <li>Bottom Section – Gene-Level Details
                    <br>
                    If you're looking for more specific insights, this section allows you to select individual genes of interest and examine their expression patterns in greater detail.
                </li>
            </ul>

            <p class="underline">
                Not sure what a graph is showing?
            </p>
            <p>
                No worries, next to each graph’s title, you’ll find a small "<sl-icon name="info-circle" class="dark-gray"></sl-icon>" icon. Simply hover over it to see helpful information explaining the graph. Also, in case your seeing a negative timepoint of -1, and you are wondering what that means. This timepoint represents the control.
            </p>

            <br>

            <p class="underline">
                Can i download the underlying expression data that is being displayed?
            </p>

            <p>Yes! At the top left of the page, you’ll find the information section for the selected study. Right next to it is a download button ("<sl-icon name="download" class="col3"></sl-icon>"). Clicking this button will download a ZIP folder containing all relevant information, including the expression tables and other dataset details.</p>

            <br>

            <p class="underline">
                Can i download the phage gene classification data?
            </p>

            <p>Yes, you can download a CSV table listing all genes and their classifications. To do so, click "<svg viewBox="0 0 857.1 1000" class="icon" height="1em" width="1em"><path d="m214-7h429v214h-429v-214z m500 0h72v500q0 8-6 21t-11 20l-157 156q-5 6-19 12t-22 5v-232q0-22-15-38t-38-16h-322q-22 0-37 16t-16 38v232h-72v-714h72v232q0 22 16 38t37 16h465q22 0 38-16t15-38v-232z m-214 518v178q0 8-5 13t-13 5h-107q-7 0-13-5t-5-13v-178q0-8 5-13t13-5h107q7 0 13 5t5 13z m357-18v-518q0-22-15-38t-38-16h-750q-23 0-38 16t-16 38v750q0 22 16 38t38 16h517q23 0 50-12t42-26l156-157q16-15 27-42t11-49z" transform="matrix(1 0 0 -1 0 850)" style="fill: rgba(68, 68, 68, 0.3);"></path></svg>" above the Phage Gene Expression Profiles plot. </p>

            <br>


            <p class="underline">How to interact with the data visualizations?</p>
            <div class="column-wrapper">
                <div class="text-wrapper">
                    
                    <p>All graphs are fully interactive, allowing you to explore the data in more detail. For more details, you can also refer to the video on the right, which demonstrates an example interaction.</p>
                    <br>
                    <p>	<strong>Hovering</strong> over any data point displays a tooltip with additional information, such as the exact expression value or other relevant details.</p>
                    <p><strong>Zoom in</strong> by clicking and dragging to create a rectangle around the area of interest. Releasing the mouse zooms into the selected region.</p>
                    <br>
                    <p>To assist with interaction, most graphs feature a set of buttons located above them, each serving a specific purpose: </p>
        
                    <ul>
                        <li><svg viewBox="0 0 1000 1000" class="icon" height="1em" width="1em"><path d="m500 450c-83 0-150-67-150-150 0-83 67-150 150-150 83 0 150 67 150 150 0 83-67 150-150 150z m400 150h-120c-16 0-34 13-39 29l-31 93c-6 15-23 28-40 28h-340c-16 0-34-13-39-28l-31-94c-6-15-23-28-40-28h-120c-55 0-100-45-100-100v-450c0-55 45-100 100-100h800c55 0 100 45 100 100v450c0 55-45 100-100 100z m-400-550c-138 0-250 112-250 250 0 138 112 250 250 250 138 0 250-112 250-250 0-138-112-250-250-250z m365 380c-19 0-35 16-35 35 0 19 16 35 35 35 19 0 35-16 35-35 0-19-16-35-35-35z" transform="matrix(1 0 0 -1 0 850)" style="fill: rgba(68, 68, 68, 0.3);"></path></svg> - Download the current graph as a PNG image.</li>
                        <li><svg viewBox="0 0 1000 1000" class="icon" height="1em" width="1em"><path d="m1000 350l-187 188 0-125-250 0 0 250 125 0-188 187-187-187 125 0 0-250-250 0 0 125-188-188 186-187 0 125 252 0 0-250-125 0 187-188 188 188-125 0 0 250 250 0 0-126 187 188z" transform="matrix(1 0 0 -1 0 850)" style="fill: rgba(68, 68, 68, 0.3);"></path></svg> - Click and drag the graph to move to different sections.</li>
                        <li><svg viewBox="0 0 875 1000" class="icon" height="1em" width="1em"><path d="m1 787l0-875 875 0 0 875-875 0z m687-500l-187 0 0-187-125 0 0 187-188 0 0 125 188 0 0 187 125 0 0-187 187 0 0-125z" transform="matrix(1 0 0 -1 0 850)" style="fill: rgba(68, 68, 68, 0.3);"></path></svg>/<svg viewBox="0 0 875 1000" class="icon" height="1em" width="1em"><path d="m0 788l0-876 875 0 0 876-875 0z m688-500l-500 0 0 125 500 0 0-125z" transform="matrix(1 0 0 -1 0 850)" style="fill: rgba(68, 68, 68, 0.3);"></path></svg> - Use these to zoom in or out incrementally.</li>
                        <li><svg viewBox="0 0 1000 1000" class="icon" height="1em" width="1em"><path d="m250 850l-187 0-63 0 0-62 0-188 63 0 0 188 187 0 0 62z m688 0l-188 0 0-62 188 0 0-188 62 0 0 188 0 62-62 0z m-875-938l0 188-63 0 0-188 0-62 63 0 187 0 0 62-187 0z m875 188l0-188-188 0 0-62 188 0 62 0 0 62 0 188-62 0z m-125 188l-1 0-93-94-156 156 156 156 92-93 2 0 0 250-250 0 0-2 93-92-156-156-156 156 94 92 0 2-250 0 0-250 0 0 93 93 157-156-157-156-93 94 0 0 0-250 250 0 0 0-94 93 156 157 156-157-93-93 0 0 250 0 0 250z" transform="matrix(1 0 0 -1 0 850)" style="fill: rgba(68, 68, 68, 0.3);"></path></svg> - Reset the zoom and pan settings to their default state.</li>
                        <li><svg viewBox="0 0 857.1 1000" class="icon" height="1em" width="1em"><path d="m500 82v107q0 8-5 13t-13 5h-107q-8 0-13-5t-5-13v-107q0-8 5-13t13-5h107q8 0 13 5t5 13z m143 375q0 49-31 91t-77 65-95 23q-136 0-207-119-9-14 4-24l74-55q4-4 10-4 9 0 14 7 30 38 48 51 19 14 48 14 27 0 48-15t21-33q0-21-11-34t-38-25q-35-16-65-48t-29-70v-20q0-8 5-13t13-5h107q8 0 13 5t5 13q0 10 12 27t30 28q18 10 28 16t25 19 25 27 16 34 7 45z m214-107q0-117-57-215t-156-156-215-58-216 58-155 156-58 215 58 215 155 156 216 58 215-58 156-156 57-215z" transform="matrix(1 0 0 -1 0 850)" style="fill: rgba(68, 68, 68, 0.3);"></path></svg> - Opens this Help page for more guidance on how to use the visualizations.</li>
                    </ul>
                </div>

                

                <div class="img-wrapper" id="heatmap-video-wrapper">
                    <video controls autoplay muted loop aria-label="Video on how to interact with a graph in dataset exploration">

                        <source src="static/assets/help/example_graph_interaction.mp4" type="video/mp4">
                                Your browser does not support the video tag.
                    </video>
                    <p class="small-p">Example: How to interact with a graph in dataset exploration.</p>
                </div>
            </div>

            <br>

            <p>In the gene selection section, you’ll also find a linear representation of the phage genome, with selected genes highlighted in color. For guidance on interacting with the linear genome views, please refer to the genome viewer guide section. The genome view in Dataset Exploration additionally provides the feature that each gene is clickable. If a gene is clicked, the respective gene will be selected and all visualizations within the gene selection area will be updated.</p>
                        


            <br>
            <sl-divider></sl-divider>
            <br>


            <!-- Genome Viewer -->
            <a href="/genome-viewer">
                <div class="heading-wrapper underline" id="guide-viewer">
                    <svg width="25" height="24" viewBox="0 0 25 24" fill="none" xmlns="http://www.w3.org/2000/svg">
                        <path d="M16.0033 19.4757L16.0026 19.4633C16.0019 19.4522 16.0008 19.4345 15.9992 19.4106C15.9958 19.3626 15.9903 19.2898 15.9819 19.1956C15.9649 19.0071 15.9361 18.7344 15.8894 18.4058C15.8666 18.2461 15.8398 18.0744 15.8083 17.8938H8.71294V16.3938H15.4668C15.4311 16.27 15.3932 16.1458 15.3529 16.0219C15.3197 15.9201 15.2853 15.8193 15.2495 15.72H8.71294V14.22H14.5176C14.4205 14.0758 14.3183 13.9414 14.2108 13.8185C13.7485 13.2899 13.2181 13 12.547 13C11.179 13 10.133 12.3618 9.37768 11.4982C8.64317 10.6584 8.16138 9.58702 7.83921 8.59689C7.51344 7.59567 7.32849 6.60665 7.22452 5.87599C7.17223 5.50849 7.1397 5.20115 7.12013 4.98362C7.11033 4.87475 7.10375 4.78806 7.09955 4.72723C7.09745 4.69681 7.09594 4.67284 7.09492 4.65576L7.09375 4.63536L7.09341 4.62914L7.09331 4.62705C7.09329 4.62673 7.09323 4.62564 8.09195 4.57495L7.09323 4.62564C7.06524 4.07406 7.48968 3.60423 8.04126 3.57624C8.59257 3.54826 9.06222 3.97228 9.09062 4.52347L9.09072 4.5253L9.09138 4.53674C9.09204 4.54776 9.09314 4.56548 9.0948 4.58944C9.09811 4.63738 9.10361 4.71022 9.11208 4.80437C9.12905 4.9929 9.15782 5.26565 9.20458 5.59424C9.2273 5.75393 9.2541 5.92561 9.2856 6.1062H17.0054V7.6062H9.62716C9.66282 7.72997 9.70075 7.85415 9.74107 7.97808C9.77421 8.07993 9.80867 8.18067 9.84447 8.28003H16.3942V9.78003H10.5763C10.6734 9.92425 10.7756 10.0586 10.8831 10.1815C11.3454 10.7101 11.8758 11 12.547 11C13.9149 11 14.9609 11.6382 15.7163 12.5018C16.4508 13.3416 16.9326 14.413 17.2547 15.4031C17.5805 16.4043 17.7655 17.3934 17.8694 18.124C17.9217 18.4915 17.9542 18.7988 17.9738 19.0164C17.9836 19.1253 17.9902 19.2119 17.9944 19.2728C17.9965 19.3032 17.998 19.3272 17.999 19.3442L18.0002 19.3646L18.0005 19.3709L18.0006 19.3729C18.0007 19.3733 18.0007 19.3744 17.002 19.425L18.0007 19.3744C18.0287 19.9259 17.6043 20.3958 17.0527 20.4238C16.5014 20.4517 16.0317 20.0269 16.0033 19.4757Z" fill="#1D1B20"></path>
                        <path d="M9.09053 19.4755L9.09059 19.4745L9.09125 19.463C9.09191 19.452 9.09302 19.4343 9.09467 19.4103C9.09798 19.3624 9.10348 19.2896 9.11195 19.1954C9.12892 19.0069 9.1577 18.7341 9.20445 18.4055C9.29859 17.744 9.46274 16.8767 9.74095 16.0217C10.0228 15.1556 10.3998 14.3707 10.883 13.8183C11.3453 13.2897 11.8758 13 12.547 13C13.9149 13 14.9608 12.3616 15.7161 11.498C16.4507 10.6582 16.9324 9.5868 17.2546 8.59667C17.5804 7.59545 17.7653 6.60643 17.8693 5.87577C17.9216 5.50827 17.9541 5.20093 17.9737 4.9834C17.9835 4.87453 17.9901 4.78784 17.9943 4.72701C17.9964 4.69659 17.9979 4.67262 17.9989 4.65554L18.0001 4.63514L18.0004 4.62892L18.0005 4.62683C18.0005 4.62651 18.0006 4.62542 17.0019 4.57473L18.0006 4.62542C18.0286 4.07384 17.6041 3.60401 17.0526 3.57602C16.5013 3.54804 16.0316 3.97206 16.0032 4.52325C16.0032 4.52327 16.0032 4.52323 16.0032 4.52325L16.0031 4.52508L16.0024 4.53652C16.0018 4.54754 16.0007 4.56526 15.999 4.58922C15.9957 4.63716 15.9902 4.71 15.9817 4.80415C15.9648 4.99268 15.936 5.26543 15.8892 5.59402C15.7951 6.25557 15.631 7.12282 15.3528 7.97786C15.0709 8.844 14.6939 9.62889 14.2107 10.1813C13.7484 10.7099 13.2181 11 12.547 11C11.179 11 10.1329 11.638 9.37755 12.5016C8.64305 13.3414 8.16125 14.4128 7.83909 15.4029C7.51331 16.4041 7.32837 17.3931 7.2244 18.1238C7.1721 18.4913 7.13958 18.7986 7.12 19.0162C7.11021 19.125 7.10363 19.2117 7.09942 19.2725C7.09732 19.303 7.09582 19.3269 7.0948 19.344L7.09362 19.3644L7.09329 19.3706L7.09318 19.3727C7.09316 19.3731 7.09311 19.3741 8.09182 19.4248L7.09311 19.3741C7.06511 19.9257 7.48956 20.3955 8.04113 20.4235C8.59244 20.4515 9.06213 20.0267 9.09053 19.4755Z" fill="#1D1B20"></path>
                    </svg>
                    <strong>Genome Viewer</strong>
                </div>
            </a>
            <br>
            <p>Explore the genome of your phage of interest labelled with phage gene classes resulting from a selected dual RNA-seq dataset.</p>
            <br>
            <p class="underline">Where to start?</p>
            <p>
                If you are interested in viewing a specific phage genome make your selections on the left side on the genome viewer page or select a specific genome and dataset by clicking the "<svg width="25" height="24" viewBox="0 0 25 24" fill="none" xmlns="http://www.w3.org/2000/svg">
                    <path d="M16.0033 19.4757L16.0026 19.4633C16.0019 19.4522 16.0008 19.4345 15.9992 19.4106C15.9958 19.3626 15.9903 19.2898 15.9819 19.1956C15.9649 19.0071 15.9361 18.7344 15.8894 18.4058C15.8666 18.2461 15.8398 18.0744 15.8083 17.8938H8.71294V16.3938H15.4668C15.4311 16.27 15.3932 16.1458 15.3529 16.0219C15.3197 15.9201 15.2853 15.8193 15.2495 15.72H8.71294V14.22H14.5176C14.4205 14.0758 14.3183 13.9414 14.2108 13.8185C13.7485 13.2899 13.2181 13 12.547 13C11.179 13 10.133 12.3618 9.37768 11.4982C8.64317 10.6584 8.16138 9.58702 7.83921 8.59689C7.51344 7.59567 7.32849 6.60665 7.22452 5.87599C7.17223 5.50849 7.1397 5.20115 7.12013 4.98362C7.11033 4.87475 7.10375 4.78806 7.09955 4.72723C7.09745 4.69681 7.09594 4.67284 7.09492 4.65576L7.09375 4.63536L7.09341 4.62914L7.09331 4.62705C7.09329 4.62673 7.09323 4.62564 8.09195 4.57495L7.09323 4.62564C7.06524 4.07406 7.48968 3.60423 8.04126 3.57624C8.59257 3.54826 9.06222 3.97228 9.09062 4.52347L9.09072 4.5253L9.09138 4.53674C9.09204 4.54776 9.09314 4.56548 9.0948 4.58944C9.09811 4.63738 9.10361 4.71022 9.11208 4.80437C9.12905 4.9929 9.15782 5.26565 9.20458 5.59424C9.2273 5.75393 9.2541 5.92561 9.2856 6.1062H17.0054V7.6062H9.62716C9.66282 7.72997 9.70075 7.85415 9.74107 7.97808C9.77421 8.07993 9.80867 8.18067 9.84447 8.28003H16.3942V9.78003H10.5763C10.6734 9.92425 10.7756 10.0586 10.8831 10.1815C11.3454 10.7101 11.8758 11 12.547 11C13.9149 11 14.9609 11.6382 15.7163 12.5018C16.4508 13.3416 16.9326 14.413 17.2547 15.4031C17.5805 16.4043 17.7655 17.3934 17.8694 18.124C17.9217 18.4915 17.9542 18.7988 17.9738 19.0164C17.9836 19.1253 17.9902 19.2119 17.9944 19.2728C17.9965 19.3032 17.998 19.3272 17.999 19.3442L18.0002 19.3646L18.0005 19.3709L18.0006 19.3729C18.0007 19.3733 18.0007 19.3744 17.002 19.425L18.0007 19.3744C18.0287 19.9259 17.6043 20.3958 17.0527 20.4238C16.5014 20.4517 16.0317 20.0269 16.0033 19.4757Z" fill="#1D1B20"></path>
                    <path d="M9.09053 19.4755L9.09059 19.4745L9.09125 19.463C9.09191 19.452 9.09302 19.4343 9.09467 19.4103C9.09798 19.3624 9.10348 19.2896 9.11195 19.1954C9.12892 19.0069 9.1577 18.7341 9.20445 18.4055C9.29859 17.744 9.46274 16.8767 9.74095 16.0217C10.0228 15.1556 10.3998 14.3707 10.883 13.8183C11.3453 13.2897 11.8758 13 12.547 13C13.9149 13 14.9608 12.3616 15.7161 11.498C16.4507 10.6582 16.9324 9.5868 17.2546 8.59667C17.5804 7.59545 17.7653 6.60643 17.8693 5.87577C17.9216 5.50827 17.9541 5.20093 17.9737 4.9834C17.9835 4.87453 17.9901 4.78784 17.9943 4.72701C17.9964 4.69659 17.9979 4.67262 17.9989 4.65554L18.0001 4.63514L18.0004 4.62892L18.0005 4.62683C18.0005 4.62651 18.0006 4.62542 17.0019 4.57473L18.0006 4.62542C18.0286 4.07384 17.6041 3.60401 17.0526 3.57602C16.5013 3.54804 16.0316 3.97206 16.0032 4.52325C16.0032 4.52327 16.0032 4.52323 16.0032 4.52325L16.0031 4.52508L16.0024 4.53652C16.0018 4.54754 16.0007 4.56526 15.999 4.58922C15.9957 4.63716 15.9902 4.71 15.9817 4.80415C15.9648 4.99268 15.936 5.26543 15.8892 5.59402C15.7951 6.25557 15.631 7.12282 15.3528 7.97786C15.0709 8.844 14.6939 9.62889 14.2107 10.1813C13.7484 10.7099 13.2181 11 12.547 11C11.179 11 10.1329 11.638 9.37755 12.5016C8.64305 13.3414 8.16125 14.4128 7.83909 15.4029C7.51331 16.4041 7.32837 17.3931 7.2244 18.1238C7.1721 18.4913 7.13958 18.7986 7.12 19.0162C7.11021 19.125 7.10363 19.2117 7.09942 19.2725C7.09732 19.303 7.09582 19.3269 7.0948 19.344L7.09362 19.3644L7.09329 19.3706L7.09318 19.3727C7.09316 19.3731 7.09311 19.3741 8.09182 19.4248L7.09311 19.3741C7.06511 19.9257 7.48956 20.3955 8.04113 20.4235C8.59244 20.4515 9.06213 20.0267 9.09053 19.4755Z" fill="#1D1B20"></path>
                </svg>" button in the dataset table on the data overview page.
            </p>
            <br>
            <p class="underline">Why do i have to choose a dataset/study</p>
            <p>This selection is necessary because the available gene classes depend on the chosen dataset. You can also customize the gene classification method by selecting a threshold, which determines how genes are classified. There are three threshold options available:</p>
            <ul>
                <li><strong>Class Max</strong>: classification is based on the gene's highest expression level.</li>
                <li><strong>Class Threshold</strong>: classification is based on a defined expression threshold of 0.2.</li>
                <li><strong>Custom Threshold</strong>: classification is based on a custom threshold. For each classification (early phase, middle phase, late phase) a time point threshold for can be selected together with a threshold between 0 and 1.</li>
            </ul>


            <p class="underline">
                What can i see in the genome visualization?
            </p>
            <p>Once all selections on the left are completed, the genome visualizations on the right will update automatically. These include a static circular view of the phage genome and interactive linear tracks that are linked to it.</p>
            <p>Currently, the genome is displayed using two tracks (or rings). In the circular view, the outer ring represents gene classification (which corresponds to the top track in the linear view), while the inner ring (bottom track in the linear view) displays gene biotypes.</p>

            <p>Each track consists of genes, that are represented as colored blocks with an arrow indicating its direction — either on the forward strand (<svg class="gene-viz" width="43" height="19" viewBox="0 0 43 19" fill="none" xmlns="http://www.w3.org/2000/svg">
                <path d="M42.1377 9.39258L33.6953 18.7402V18.7246H0.43457V0.0605469H33.6953V0.0449219L42.1377 9.39258Z" fill="#DA8439"/>
                </svg>) or reverse strand (<svg class="gene-viz" width="43" height="19" viewBox="0 0 43 19" fill="none" xmlns="http://www.w3.org/2000/svg">
                    <path d="M0.43457 9.49512L8.87695 0.147461V0.163086H42.1377V18.8271H8.87695V18.8428L0.43457 9.49512Z" fill="#DA8439"/></svg>). To know which color represents what, please refer to the legend that is located on the right side of the linear tracks.
            </p>

            <br>
            <p class="underline">
                How to navigate the genome visualization?
            </p>

            <p>Circular View:</p>
            <ul>
                <li>The circular view of the genome remains static, meaning you cannot zoom in or out or scroll within it. It serves as a high-level overview of the genome.</li>
                <li>A gray window is located within the circular view. This window is used to define the region of the genome being displayed in the linear tracks.</li>
                <li>Dragging the gray window across the circular view will update the linear tracks accordingly. The linear tracks will show more detailed data for the region defined by this window. It is also possible to manually expand the window in the circular view which will then update the linear tracks accordingly.</li>
            </ul>
            
               
            <p>Linear Tracks:</p>

            <ul>
                <li>The linear tracks on the right side of the viewer are fully interactive, allowing you to zoom in and out for more details or broader views of the genome.</li>
                <li>Zooming in will focus on a smaller portion of the genome, providing more detail. Zooming out will show a broader stretch of the genome, and the corresponding gray window on the circular view will grow in size to represent the expanded view.</li>
                <li>Scrolling through the linear tracks allows you to explore different sections of the genome, letting you move left and right across the dataset.</li>
            </ul>
                
            <p>Detailed Navigation and Interactivity:</p>
            <div class="column-wrapper">
                <div class="text-wrapper">
                    <ol>
                        <li>Navigating with the Gray Window:
                            <ul>
                                <li>Drag the gray window within the circular view to update the corresponding region in the linear tracks by clicking on the window and dragging it.</li>
                                <li>Change the size of the gray window by hovering over its border until the "<svg class="black" width="58" height="57" viewBox="0 0 58 57" fill="none" xmlns="http://www.w3.org/2000/svg">
                                    <path d="M57.2498 28.3269L48.7156 36.1511V32.0183H32.6394V48.0095H36.7722L28.948 56.5437L21.1238 48.0095H25.2566V32.0183H9.25952V36.1511L0.726318 28.3269L9.25952 20.5027V24.6355H48.7156V20.5027L57.2498 28.3269ZM36.7722 8.55347H32.6394V24.6326H25.2566V8.55347H21.1238L28.948 0.0202637L36.7722 8.55347Z" fill="black"/>
                                </svg>" cursor appears. If it appears, click and drag the window to the prefered size. The linear view will update accordingly. 
                                </li>
                                <li>The size of the gray window will also change based on the zoom level of the linear tracks, giving you a visual representation of the region you’re focusing on.</li>
                            </ul>
                        </li>
                            
                        <li>Zooming in and Out:
                            <ul>
                                <li>Zooming in in the linear tracks gives you more detail for the selected region, while zooming out expands the region. This is done by hovering over your area of interest on the linear tracks and scrolling in with either your mouse or trackpad.</li>
                                <li>The views in the circular and linear tracks are linked, meaning any zoom action on the linear tracks is reflected in the circular view.</li>
                            </ul>
                        </li>
                            
                        <li>Scrolling:
                            <ul>
                                <li>Scrolling through the genome can be done either by dragging the gray window over the circular view or by clicking and dragging on the linear tracks</li>
                            </ul>
                        </li>
                            

                        <li>Hovering for Details
                            <ul>
                                <li>Hover over genes on either the circular genome or the linear tracks to reveal additional details in the tooltip.</li>
                                <li>This gives you a quick way to gather information about each gene, such as its location, biotype, and strand.</li>
                            </ul>
                        </li>
                            
                    </ol>  
                </div>


                <div class="img-wrapper">
                    <video controls autoplay muted loop aria-label="Video on how to navigate the genome visualization">
                        <source src="static/assets/help/example_genome.mp4" type="video/mp4">
                        Your browser does not support the video tag.
                    </video>
                    <p class="small-p">Example: How to navigate the genome visualization.</p>
                </div>

                

            </div>

        </sl-details>

        <sl-details summary="Terms of Use" open id="terms-of-use">
            This website is intended to provide access to and enable visualization of dual RNA-seq data of phage infections. By using this platform, you agree to the following terms:
                <ul>
                    <li>All content is provided for informational and academic purposes only.</li>
                    <li>You may access, use, and cite data freely, provided appropriate credit is given to the source and publication.</li>
                    <li>Commercial use of the content is not permitted without prior written consent.</li>
                
                </ul>
                We reserve the right to modify or discontinue the platform.
                For citation and reuse, please refer to the associated publication of the PhageExpressionAtlas (<em>Link to be added</em>) as well as original studies that your datasets of interest are derived from.
                <!-- TODO insert puclication link -->
        </sl-details>

        <!-- TODO: Github link and reach out link  -->
        <sl-details summary="Impressum (Legal Disclosure)" open id="impressum">
            <u>Contact:</u> <br>
            Dr. Maik Wolfram-Schauerte <br>
            <a href="mailto:maik.wolfram-schauerte@uni-tuebingen.de">maik.wolfram-schauerte@uni-tuebingen.de</a> <br>
            University of Tübingen, Institute for Bioinformatics and Medical Informatics <br>
            Sand 14, D-72076 Tübingen 
        </sl-details>

        <!-- TODO: Links ergänzen -->
        <sl-details summary="Contribution & Feedback" open id="feedback">
            Do you have dual RNA-seq data of a phage infection that is not yet included in the PhageExpressionAtlas?<br>
            You can <a href="">let us know</a> and we will process your data and include it in the atlas. Alternatively, you can process that data yourself via our <a href="">pipeline</a> and send the counts table directly to us.

            <br>

            You are interested in how data is preprocessed and stored or how we implemented the frontend of the PhageExpressionAtlas? <br>
            Check out our <a href="">GitHub</a> or feel free to <a href="mailto:maik.wolfram-schauerte@uni-tuebingen.de">reach out to us</a>. We also welcome report of bugs, feedback on general impressions or suggestions for optimization. Thanks for helping to improve the PhageExpressionAtlas.
        </sl-details>

        <sl-details summary="Privacy Policy" open id="privacy-policy">
            We take the protection of your personal data seriously. This website complies with the EU General Data Protection Regulation (GDPR) and applicable German data protection laws.
            <br>
            <br>
            <strong> Data Collection and Use</strong> <br>
            We do not collect personal data unless you voluntarily provide it (e.g., via email contact). The data will only be used for the purpose for which it was provided and will not be shared with third parties without your consent.
            <br>
            <br>
            <strong> Server Logs</strong> <br>
            When accessing this website, your browser automatically sends information to our web server. This includes:
            <ul>
                <li>IP address</li>
                <li>Date and time of access</li>
                <li>Browser type and version</li>
                <li>Operating system</li>
            </ul>
            This data is stored temporarily in server log files for technical reasons and to ensure the stability and security of the website. It is not merged with other data sources.
            <br>
            <br>
            <strong>Cookies</strong> <br>
            This website does not use cookies for tracking or advertising purposes. If cookies are used (e.g., for session handling), they are strictly necessary for functionality and are not used to personally identify users.
            <br>
            <br>
            <strong>Contact</strong> <br>
            If you contact us via email, we will store your message and the contact information you provide for the purpose of responding to your inquiry. This data will not be shared.
            <br>
            <br>
            <strong>Data Subject Rights</strong> <br>
            You have the right to:
            <ul>
                <li>Access your stored data</li>
                <li>Request correction or deletion</li>
                <li>Object to data processing</li>
                <li>File a complaint with the responsible data protection authority</li>
            </ul>    
            For any data protection inquiries, please contact us at: <a href="mailto:maik.wolfram-schauerte@uni-tuebingen.de">maik.wolfram-schauerte@uni-tuebingen.de</a>
        </sl-details>

        <sl-details summary="Disclaimer" open id="disclaimer">
            The content provided on this website has been compiled with the greatest care and is based on the best available scientific knowledge. However, we do not guarantee the accuracy or completeness of the provided datasets.
            <br>
            We assume no liability for any direct or indirect damages resulting from the use of this platform, including decisions made or actions taken based on the data provided. Users are responsible for critically evaluating the content and, where appropriate, referring to the original sources.
            <br>
            For details on data collection, curation, processing, and visualization methods, please refer to the associated publication: <em>Link to be added</em>
            <!--TODO: ADD LINK OF PAPER -->
        </sl-details>

        <sl-details summary="License" open id="license">
            The PhageExpressionAtlas is available under the terms of the MIT license.
        </sl-details>



        <sl-details summary="Accessibility Statement" open id="accessibility-statement">
            
            We are committed to making this website accessible in accordance with national legislation implementing Directive (EU) 2016/2102.<br>
            <br>
            <strong>Current Status of Accessibility:</strong>
            <br>
            This website is partially compliant with the applicable accessibility standards. We are working to improve accessibility features such as:</strong>
            <ul>
                <li>Text alternatives for non-text content</li>
                <li>Contrast and readability improvements</li>
            </ul>
            If you encounter accessibility barriers while using this site, please contact us so we can address the issue: <a href="mailto:maik.wolfram-schauerte@uni-tuebingen.de">maik.wolfram-schauerte@uni-tuebingen.de</a>.
        </sl-details>

        <sl-details summary="Acknowledgement" open>
            The PhageExpressionAtlas has been created and is maintained by the Integrative Transcriptomics group at the Institute for Bioinformatics and Medical Informatics (IBMI) at Tübingen University. The atlas is hosted on the IBMI server infrastructure and part of <a href="https://tuevis.cs.uni-tuebingen.de">TueVis</a>. Work on the PhageExpressionAtlas is supported by the Wilhelm Schuler-Stiftung (Tübingen University).
            <br>
            <br>
            <strong>Current contributors</strong>
            <ul>
                <li><strong>Dr. Maik Wolfram-Schauerte</strong> (project lead, backend, conceptualization, supervision, funding acquisition)</li>
                <li><strong>Caroline Trust</strong> (design, frontend, conceptualization)</li>
                <li><strong>Nils Waffenschmidt</strong> (backend)</li>
                <li><strong>Prof. Dr. Kay Nieselt</strong> (supervision, funding acquisition)</li>
            </ul>
            
            Moreover, we want to acknowledge all authors from the various studies that generated data included in this atlas. Your contributions were invaluable for this resource.
            
        </sl-details>

        <sl-details summary="Attributions" open>
            <ul id="ul-margin">
                <li><a href="https://shoelace.style/">Shoelace</a>
                    <ul>
                        <li>Description: Library of web components.</li>
                        <li>License: MIT</li>
                        <li>Version: 2.20.0</li>
                        <li>CDN provider: jsdelivr.net </li>
                    </ul>
                </li>

                <li><a href="https://echarts.apache.org/en/">ECharts</a>
                    <ul>
                        <li>Description: Library for charting and visualization.</li>
                        <li>License: Apache License 2.0</li>
                        <li>Version: 5.4.0</li>
                        <li>CDN provider: jsdelivr.net </li>
                    </ul>
                </li>

                <li><a href="https://github.com/ecomfe/echarts-wordcloud">ECharts Wordcloud</a>
                    <ul>
                        <li>Description: Word Cloud Extension based on ECharts and wordcloud2.js </li>
                        <li>License: Apache License 2.0</li>
                        <li>Version: 2.1.0</li>
                        <li>CDN provider: jsdelivr.net </li>
                    </ul>
                </li>

                <li><a href="https://plotly.com/javascript/">Plotly.js</a>
                    <ul>
                        <li>Description: Library for charting.</li>
                        <li>License: MIT</li>
                        <li>Version: 3.0.0</li>
                    </ul>
                </li>

                <li><a href="https://gosling-lang.org/">Gosling</a>
                    <ul>
                        <li>Description: Toolkit for genomics data visualization</li>
                        <li>License: MIT</li>
                        <li>Version: 1.0.0</li>
                    </ul>
                </li>

                <li><a href="https://tabulator.info/">Tabulator</a>
                    <ul>
                        <li>Description: A JavaScript library for creating interactive data tables.</li>
                        <li>License: MIT</li>
                        <li>Version: 6.3.0</li>
                        <li>CDN provider: unpgk.com </li>
                    </ul>
                </li>

                <li><a href="https://axios-http.com/">Axios</a>
                    <ul>
                        <li>Description: A promise-based HTTP client for making requests in JavaScript.</li>
                        <li>License: MIT</li>
                        <li>Version: 1.8.4</li>
                        <li>CDN provider: jsdelivr.net </li>
                    </ul>
                </li>

                <li><a href="https://flask.palletsprojects.com/en/stable/">Flask</a>
                    <ul>
                        <li>Description: A lightweight WSGI web application framework.</li>
                        <li>License: BSD-3-Clause License</li>
                        <li>Version: 3.1.0</li>
                    </ul>
                </li>

                <li><a href="https://www.papaparse.com/faq">Papaparse</a>
                    <ul>
                        <li>Description: CSV Parser.</li>
                        <li>License: MIT</li>
                        <li>Version: 5.5.2</li>
                    </ul>
                </li>

                <li><a href="https://stuk.github.io/jszip/">JSZIP</a>
                    <ul>
                        <li>Description: Library for creating, reading and editing zip files.</li>
                        <li>License: MIT</li>
                        <li>Version: 3.10.1</li>
                    </ul>
                </li>
                
                <li><a href="https://github.com/guybedford/es-module-shims#readme">Es Module Shims</a>
                    <ul>
                        <li>Description: Tool that polyfills import maps and other ES Modules .</li>
                        <li>License: MIT</li>
                        <li>Version: 2.0.10</li>
                    </ul>
                </li>
    
                <li><a href="https://unpkg.com/">unpkg (for CDN management)</a>
                    <ul>
                        <li>Description: A fast, global content delivery network for everything on npm.</li>
                    </ul>
                </li>
                
                <li><a href="https://www.jsdelivr.com/">jsdelivr (for CDN management)</a>
                    <ul>
                        <li>Description: A free CDN that serves JavaScript libraries, CSS frameworks, and other assets with high speed and reliability.</li>
                    </ul>
                </li>
                
                
            </ul>

        </sl-details>

        <sl-details summary="Features to come in future" open>
            <!--Removed intentionally-->
            <!-- <ul>
                <li>analysis pipeline for and visualization of dual differential RNA-seq and ONT-cappable-seq data of phage infections to investigate transcriptome architecture of phage and host</li>
                <li>cross-phage and cross-host analyses including differential expression analysis to compare similar and distantly related phages and hosts</li>
                <li>beyond dual RNA-seq data: triple, quadruple, … RNA-seq including several phages or hosts sequenced within one sample</li>
                <li>searching for (potential) functions of your gene of interest</li>
            </ul> -->

            Are there more and not yet available functions you are interested in using in the PhageExpressionAtlas? <br>
            <a href="mailto:maik.wolfram-schauerte@uni-tuebingen.de">Please let us know!</a>

        </sl-details>

    
    </div>



</div>

{% endblock %}
